<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Clarice Groeneveld">

<title>R and Seurat to analyse single cell RNA-seq – Blog | Clarice S Groeneveld</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog | Clarice S Groeneveld</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/csgroen"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/cs_groen"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">R and <strong>Seurat</strong> to analyse single cell RNA-seq</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
            <p class="subtitle lead">Case study - NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Clarice Groeneveld </p>
            </div>
    </div>
      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#pre-requisites" id="toc-pre-requisites" class="nav-link active" data-scroll-target="#pre-requisites">Pre-requisites</a></li>
  <li><a href="#resources" id="toc-resources" class="nav-link" data-scroll-target="#resources">Resources</a></li>
  <li><a href="#install-packages" id="toc-install-packages" class="nav-link" data-scroll-target="#install-packages">Install packages</a></li>
  <li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data">Download data</a></li>
  <li><a href="#read-10x-data" id="toc-read-10x-data" class="nav-link" data-scroll-target="#read-10x-data">Read 10X data</a>
  <ul>
  <li><a href="#merge-samples" id="toc-merge-samples" class="nav-link" data-scroll-target="#merge-samples">Merge samples</a></li>
  </ul></li>
  <li><a href="#seurat-syntax-basics-for-object-accession" id="toc-seurat-syntax-basics-for-object-accession" class="nav-link" data-scroll-target="#seurat-syntax-basics-for-object-accession">Seurat syntax basics for object accession</a>
  <ul>
  <li><a href="#cell-metadata" id="toc-cell-metadata" class="nav-link" data-scroll-target="#cell-metadata">Cell metadata</a></li>
  <li><a href="#assay-matrix-e.g.-counts" id="toc-assay-matrix-e.g.-counts" class="nav-link" data-scroll-target="#assay-matrix-e.g.-counts">Assay matrix (e.g.&nbsp;counts)</a></li>
  <li><a href="#umi-ids" id="toc-umi-ids" class="nav-link" data-scroll-target="#umi-ids">UMI ids</a></li>
  <li><a href="#features" id="toc-features" class="nav-link" data-scroll-target="#features">Features</a></li>
  </ul></li>
  <li><a href="#pre-processing-workflow" id="toc-pre-processing-workflow" class="nav-link" data-scroll-target="#pre-processing-workflow">Pre-processing workflow</a>
  <ul>
  <li><a href="#qc-and-selecting-cells-for-further-analysis" id="toc-qc-and-selecting-cells-for-further-analysis" class="nav-link" data-scroll-target="#qc-and-selecting-cells-for-further-analysis">QC and selecting cells for further analysis</a></li>
  <li><a href="#filtering-and-normalization" id="toc-filtering-and-normalization" class="nav-link" data-scroll-target="#filtering-and-normalization">Filtering and Normalization</a></li>
  <li><a href="#feature-selection-and-pca" id="toc-feature-selection-and-pca" class="nav-link" data-scroll-target="#feature-selection-and-pca">Feature selection and PCA</a>
  <ul class="collapse">
  <li><a href="#dimensionality-determination" id="toc-dimensionality-determination" class="nav-link" data-scroll-target="#dimensionality-determination">Dimensionality determination</a></li>
  </ul></li>
  <li><a href="#doublet-detection" id="toc-doublet-detection" class="nav-link" data-scroll-target="#doublet-detection">Doublet detection</a></li>
  </ul></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">Clustering</a></li>
  <li><a href="#integration" id="toc-integration" class="nav-link" data-scroll-target="#integration">Integration</a>
  <ul>
  <li><a href="#check-biases" id="toc-check-biases" class="nav-link" data-scroll-target="#check-biases">Check biases</a></li>
  </ul></li>
  <li><a href="#annotation" id="toc-annotation" class="nav-link" data-scroll-target="#annotation">Annotation</a>
  <ul>
  <li><a href="#differential-expression" id="toc-differential-expression" class="nav-link" data-scroll-target="#differential-expression">Differential Expression</a>
  <ul class="collapse">
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap">Heatmap</a></li>
  <li><a href="#dot-plot" id="toc-dot-plot" class="nav-link" data-scroll-target="#dot-plot">Dot plot</a></li>
  <li><a href="#featureplot" id="toc-featureplot" class="nav-link" data-scroll-target="#featureplot">FeaturePlot</a></li>
  </ul></li>
  <li><a href="#reference-based" id="toc-reference-based" class="nav-link" data-scroll-target="#reference-based">Reference-based</a>
  <ul class="collapse">
  <li><a href="#built-in-references" id="toc-built-in-references" class="nav-link" data-scroll-target="#built-in-references">“Built-in” references</a></li>
  <li><a href="#single-cell" id="toc-single-cell" class="nav-link" data-scroll-target="#single-cell">Single-cell</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#extras" id="toc-extras" class="nav-link" data-scroll-target="#extras">Extras</a>
  <ul>
  <li><a href="#sub-cluster" id="toc-sub-cluster" class="nav-link" data-scroll-target="#sub-cluster">Sub-cluster</a></li>
  <li><a href="#trajectory-analysis" id="toc-trajectory-analysis" class="nav-link" data-scroll-target="#trajectory-analysis">Trajectory analysis</a></li>
  <li><a href="#cell-cycle" id="toc-cell-cycle" class="nav-link" data-scroll-target="#cell-cycle">Cell cycle</a></li>
  <li><a href="#rna-velocity" id="toc-rna-velocity" class="nav-link" data-scroll-target="#rna-velocity">RNA velocity</a></li>
  <li><a href="#pathway-enrichment" id="toc-pathway-enrichment" class="nav-link" data-scroll-target="#pathway-enrichment">Pathway enrichment</a></li>
  <li><a href="#gene-regulatory-network" id="toc-gene-regulatory-network" class="nav-link" data-scroll-target="#gene-regulatory-network">Gene regulatory network</a></li>
  <li><a href="#cnv-analysis" id="toc-cnv-analysis" class="nav-link" data-scroll-target="#cnv-analysis">CNV analysis</a></li>
  <li><a href="#ligand-receptor-analysis" id="toc-ligand-receptor-analysis" class="nav-link" data-scroll-target="#ligand-receptor-analysis">Ligand-receptor analysis</a></li>
  <li><a href="#object-conversion" id="toc-object-conversion" class="nav-link" data-scroll-target="#object-conversion">Object conversion</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>Download this notebook: https://raw.githubusercontent.com/csgroen/teaching/refs/heads/main/20250214_Tutorial_DU_Seurat.qmd</p>
<section id="pre-requisites" class="level2">
<h2 class="anchored" data-anchor-id="pre-requisites">Pre-requisites</h2>
<p>For the best understanding of this tutorial, please make sure you have at least an intermediate understanding of R. Here are some recommended resources to learn the basics and necessary intermediate concepts:</p>
<ol type="1">
<li>swirl: https://swirlstats.com/students.html</li>
</ol>
<p>Specifically courses <em>R Programming</em>, <em>Getting and Cleaning Data</em> and <em>Exploratory Data Analysis</em> may provide a good jump start.</p>
<ol start="2" type="1">
<li>Hands On Programming with R: https://rstudio-education.github.io/hopr/</li>
</ol>
<p>The very basics of R programming as also explained in this thorough book from the Posit team (that builds RStudio).</p>
<ol start="3" type="1">
<li>R for Data Science: https://r4ds.hadley.nz/</li>
</ol>
<p>This goes further than we need, but a base for most of the programming questions a data analyst working in R could have to start with.</p>
</section>
<section id="resources" class="level2">
<h2 class="anchored" data-anchor-id="resources">Resources</h2>
<p>Seurat has many very useful tutorials for all of its features. Elements of this case study were taken from the following:</p>
<ol type="1">
<li>Seurat - Guided Clustering Tutorial https://satijalab.org/seurat/archive/v3.0/pbmc3k_tutorial.html</li>
<li>Introduction to scRNA-seq integration: https://satijalab.org/seurat/articles/integration_introduction</li>
<li>Mapping and annotating query datasets: https://satijalab.org/seurat/articles/integration_mapping</li>
</ol>
</section>
<section id="install-packages" class="level2">
<h2 class="anchored" data-anchor-id="install-packages">Install packages</h2>
<p>To ensure we have all installed libraries, we’ll use R package “pacman”, which allows us to check if packages are not installed and load them. This is not a replacement for more comprehensive version control (e.g.&nbsp;using <code>renv</code>), but it’s simple to install and very lightweight.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span> <span class="st">"pacman"</span> <span class="sc">%in%</span> <span class="fu">installed.packages</span>()) <span class="fu">install.packages</span>(<span class="st">"pacman"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>pacman<span class="sc">::</span><span class="fu">p_load</span>(<span class="st">"Seurat"</span>, <span class="st">"tidyverse"</span>, <span class="st">"R.utils"</span>, <span class="st">"ggpubr"</span>, <span class="st">"patchwork"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>               <span class="st">"celldex"</span>, <span class="st">"SingleR"</span>, <span class="st">"scRNAseq"</span>, <span class="st">"pheatmap"</span>,</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>               <span class="st">"ggsci"</span>, <span class="st">"scds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>I’m personally not a big fan of the standard look of Seurat plots, so I’ll apply this ggplot2 theme to (almost) all Seurat plots we generate:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>theme_clean <span class="ot">&lt;-</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<blockquote class="blockquote">
<p>Note: Seurat plots are ggplot-based and therefore can be “added to” in a grammar-of-graphics manner. Plots for multiple features are patchworks, which use a slightly different syntax to add to (this is why sometimes you’ll see <code>&amp;</code> instead of <code>+</code> being used).</p>
</blockquote>
</section>
<section id="download-data" class="level2">
<h2 class="anchored" data-anchor-id="download-data">Download data</h2>
<p>Our dataset comes from Salomé et al (2022) [https://www.sciencedirect.com/science/article/pii/S1535610822003695], “NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer”. This single-cell data is a bit particular: they have FACS-sorted their cells into CD45+ and CD45-, notably to enrich their dataset in hematopoetic cells.</p>
<p>We’ll download the data from the link provided in the article:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://prod-dcd-datasets-cache-zipfiles.s3.eu-west-1.amazonaws.com/7yb7s9769c-1.zip"</span>, <span class="st">'data/Salome_2022.zip'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(<span class="st">'data/Salome_2022.zip'</span>, <span class="at">exdir =</span> <span class="st">"data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="read-10x-data" class="level2">
<h2 class="anchored" data-anchor-id="read-10x-data">Read 10X data</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ref_dir <span class="ot">&lt;-</span> <span class="st">"data/NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer/"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(ref_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "385_Patient49_CD45pos" "394_Patient50_CD45pos" "397_Patient51_CD45pos"
 [4] "421_Patient52_CD45pos" "434_Patient53_CD45pos" "435_Patient54_CD45pos"
 [7] "453_Patient01_CD45pos" "454_Patient01_CD45neg" "456_Patient02_CD45pos"
[10] "457_Patient02_CD45neg" "584_Patient03_bulk"    "595_Patient04_bulk"   
[13] "613_Patient55_bulk"    "651_Patient05_bulk"    "656_Patient06_bulk"   
[16] "663_Patient07_bulk"    "695_Patient08_bulk"   </code></pre>
</div>
</div>
<p>We can see the data is distributed as one folder per dataset, and includes the bulk data, which we’re not going to use. For this reason, I will filter out the “bulk” folders and to get my vector of <code>sc_dts</code> (single-cell datasets).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sc_dts <span class="ot">&lt;-</span> ref_dir <span class="sc">|&gt;</span> <span class="fu">list.files</span>() <span class="sc">|&gt;</span> <span class="fu">str_subset</span>(<span class="st">"bulk"</span>, <span class="at">negate =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sc_dts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "385_Patient49_CD45pos" "394_Patient50_CD45pos" "397_Patient51_CD45pos"
 [4] "421_Patient52_CD45pos" "434_Patient53_CD45pos" "435_Patient54_CD45pos"
 [7] "453_Patient01_CD45pos" "454_Patient01_CD45neg" "456_Patient02_CD45pos"
[10] "457_Patient02_CD45neg"</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Syntax note:</strong> <code>ref_dir |&gt; list.files() |&gt; str_subset("bulk", negate = TRUE)</code> is equivalent to <code>str_subset(list.files(ref_dir), "bulk", negate = TRUE)</code>. The <code>|&gt;</code> was first introduced as <code>%&gt;%</code> from R package magrittr and officially incorporated into base R&gt;=4.0. It’s purpose is to increase code legibility through “chaining” commands instead of “nesting” them. If you would like to learn more, you can check the help file <code>? |&gt;</code></p>
</blockquote>
<p>We can check inside each folder:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ref_dir <span class="sc">|&gt;</span> <span class="fu">filePath</span>(sc_dts[<span class="dv">1</span>]) <span class="sc">|&gt;</span> <span class="fu">list.files</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"  </code></pre>
</div>
</div>
<p>This is a common export of single-cell data and can be read directly into Seurat using function <code>Read10X</code>.</p>
<p>This snippet will read all of our samples at once:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>read_exp_sample <span class="ot">&lt;-</span> <span class="cf">function</span>(sc_file) {</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">'Reading file: '</span>, sc_file)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  so <span class="ot">&lt;-</span> ref_dir <span class="sc">|&gt;</span> <span class="fu">filePath</span>(sc_file) <span class="sc">|&gt;</span> <span class="fu">Read10X</span>()</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  so <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(so, <span class="at">project =</span> <span class="st">"BLCA_NKG2A"</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"sample"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(sc_file, <span class="st">"^.{4}"</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"patient"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(sc_file, <span class="st">"Patient.."</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"CD45"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(sc_file, <span class="st">"(?&lt;=CD45).*"</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(so)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sc_dts, read_exp_sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 385_Patient49_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 394_Patient50_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 397_Patient51_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 421_Patient52_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 434_Patient53_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 435_Patient54_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 453_Patient01_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 454_Patient01_CD45neg</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 456_Patient02_CD45pos</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading file: 457_Patient02_CD45neg</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sc_data) <span class="ot">&lt;-</span> sc_dts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="merge-samples" class="level3">
<h3 class="anchored" data-anchor-id="merge-samples">Merge samples</h3>
<p>The script we ran generated 10 Seurat objects, which we will merge into one for further analysis. Each count matrix, at first, will be stored as a <code>Layer</code> in our Seurat object.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(sc_data[[<span class="dv">1</span>]], sc_data[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(sc_data)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Some cell names are duplicated across objects provided. Renaming to
enforce unique cell names.</code></pre>
</div>
</div>
</section>
</section>
<section id="seurat-syntax-basics-for-object-accession" class="level2">
<h2 class="anchored" data-anchor-id="seurat-syntax-basics-for-object-accession">Seurat syntax basics for object accession</h2>
<p>Here are some “essential commands” for accession of information inside the Seurat object :<a href="https://satijalab.org/seurat/articles/essential_commands.html" class="uri">https://satijalab.org/seurat/articles/essential_commands.html</a>.</p>
<p>The document also outlines the “basic pipeline”.</p>
<p>To see some general info about the object, we can just print it:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sc_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
33538 features across 50940 samples within 1 assay 
Active assay: RNA (33538 features, 0 variable features)
 10 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10</code></pre>
</div>
</div>
<p>Note the layers, for the different samples. We can also see what analyses have been run in this object (none yet).</p>
<section id="cell-metadata" class="level4">
<h4 class="anchored" data-anchor-id="cell-metadata">Cell metadata</h4>
<p><code>[[]]</code> allows us to access the cell metadata in a concise manner:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sc_data[[]] <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                     orig.ident nCount_RNA nFeature_RNA            sample
AAACCTGAGGCCCTCA-1_1 BLCA_NKG2A       4917         1671 Patient49_CD45pos
AAACCTGAGTTCGCGC-1_1 BLCA_NKG2A        891          369 Patient49_CD45pos
AAACCTGCACTGTTAG-1_1 BLCA_NKG2A       1121          606 Patient49_CD45pos
AAACCTGCAGACGCTC-1_1 BLCA_NKG2A       3877         1484 Patient49_CD45pos
AAACCTGGTACAGCAG-1_1 BLCA_NKG2A       1393          552 Patient49_CD45pos
AAACCTGGTCACACGC-1_1 BLCA_NKG2A       1491          709 Patient49_CD45pos
                       patient CD45
AAACCTGAGGCCCTCA-1_1 Patient49  pos
AAACCTGAGTTCGCGC-1_1 Patient49  pos
AAACCTGCACTGTTAG-1_1 Patient49  pos
AAACCTGCAGACGCTC-1_1 Patient49  pos
AAACCTGGTACAGCAG-1_1 Patient49  pos
AAACCTGGTCACACGC-1_1 Patient49  pos</code></pre>
</div>
</div>
</section>
<section id="assay-matrix-e.g.-counts" class="level4">
<h4 class="anchored" data-anchor-id="assay-matrix-e.g.-counts">Assay matrix (e.g.&nbsp;counts)</h4>
<p>To get matrices from the Seurat objects, there are many “official” ways, but one of these two, which have exactly the same output, should work well:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GetAssayData</span>(sc_data, <span class="at">layer =</span> <span class="st">"counts.1"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>5 x 5 sparse Matrix of class "dgCMatrix"
            AAACCTGAGGCCCTCA-1_1 AAACCTGAGTTCGCGC-1_1 AAACCTGCACTGTTAG-1_1
MIR1302-2HG                    .                    .                    .
FAM138A                        .                    .                    .
OR4F5                          .                    .                    .
AL627309.1                     .                    .                    .
AL627309.3                     .                    .                    .
            AAACCTGCAGACGCTC-1_1 AAACCTGGTACAGCAG-1_1
MIR1302-2HG                    .                    .
FAM138A                        .                    .
OR4F5                          .                    .
AL627309.1                     .                    .
AL627309.3                     .                    .</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc_data[[<span class="st">"RNA"</span>]]<span class="sc">$</span>counts<span class="fl">.1</span>[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>5 x 5 sparse Matrix of class "dgCMatrix"
            AAACCTGAGGCCCTCA-1_1 AAACCTGAGTTCGCGC-1_1 AAACCTGCACTGTTAG-1_1
MIR1302-2HG                    .                    .                    .
FAM138A                        .                    .                    .
OR4F5                          .                    .                    .
AL627309.1                     .                    .                    .
AL627309.3                     .                    .                    .
            AAACCTGCAGACGCTC-1_1 AAACCTGGTACAGCAG-1_1
MIR1302-2HG                    .                    .
FAM138A                        .                    .
OR4F5                          .                    .
AL627309.1                     .                    .
AL627309.3                     .                    .</code></pre>
</div>
</div>
</section>
<section id="umi-ids" class="level3">
<h3 class="anchored" data-anchor-id="umi-ids">UMI ids</h3>
<p>To get the “ids” of the cells, we can simply call:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Cells</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAACCTGAGGCCCTCA-1_1" "AAACCTGAGTTCGCGC-1_1" "AAACCTGCACTGTTAG-1_1"
[4] "AAACCTGCAGACGCTC-1_1" "AAACCTGGTACAGCAG-1_1"</code></pre>
</div>
</div>
</section>
<section id="features" class="level3">
<h3 class="anchored" data-anchor-id="features">Features</h3>
<p>For a single-cell RNA-seq object, we can see the genes (features) like so:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Features</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "MIR1302-2HG" "FAM138A"     "OR4F5"       "AL627309.1"  "AL627309.3" </code></pre>
</div>
</div>
</section>
</section>
<section id="pre-processing-workflow" class="level2">
<h2 class="anchored" data-anchor-id="pre-processing-workflow">Pre-processing workflow</h2>
<p>There are very standard steps for pre-processing, but that require some level of decision on the part of the user. The basic tutorial of Seurat, on PBMCs, covers this in detail: https://satijalab.org/seurat/articles/pbmc3k_tutorial</p>
<section id="qc-and-selecting-cells-for-further-analysis" class="level3">
<h3 class="anchored" data-anchor-id="qc-and-selecting-cells-for-further-analysis">QC and selecting cells for further analysis</h3>
<p>This command will figure out the percentage of reads of mitchondrial RNA:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sc_data[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_data, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>There are three major pieces of information we’d like to check to make filtering decisions: number of expressed features, number of reads and the percentage of mitochondrial RNA. Here, I plot per sample:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>(<span class="fu">VlnPlot</span>(sc_data, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>), </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"sample"</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">alpha =</span> <span class="fl">0.3</span>, </span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">&amp;</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  theme_clean) <span class="sc">+</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-15-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="intro_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>We can also plot this for a particular known variable. For example, here I show for whether the cells were FACS-sorted as CD45 positive or negative:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>(<span class="fu">VlnPlot</span>(sc_data, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>), </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">alpha =</span> <span class="fl">0.3</span>,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">group.by =</span> <span class="st">"CD45"</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">&amp;</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_compare_means</span>() <span class="sc">&amp;</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  theme_clean) <span class="sc">+</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">+</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">"CD45"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="intro_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="filtering-and-normalization" class="level3">
<h3 class="anchored" data-anchor-id="filtering-and-normalization">Filtering and Normalization</h3>
<p>We’ll reproduce the filtering and normalization steps proposed by the paper as closely as possible:</p>
<blockquote class="blockquote">
<p>For each sample, cells were first selected as expressing less than 16–20% mitochondrial genes and displaying a minimum of 200–300 and a maximum of 2500–3500 features.</p>
</blockquote>
<p>As we’re standardizing pre-processing for all samples, we’ll take the least stringent:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">subset</span>(sc_data, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span> <span class="sc">&amp;</span> nFeature_RNA <span class="sc">&lt;</span> <span class="dv">3500</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<blockquote class="blockquote">
<p>Data were then log-normalized using a scale factor of 10,000.</p>
</blockquote>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sc_data, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="fl">1e4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.9</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts.10</code></pre>
</div>
</div>
</section>
<section id="feature-selection-and-pca" class="level3">
<h3 class="anchored" data-anchor-id="feature-selection-and-pca">Feature selection and PCA</h3>
<blockquote class="blockquote">
<p>The 2,000 most variable features were then identified, data were scaled based on all the features […]</p>
</blockquote>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(sc_data, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.9</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts.10</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(sc_data, <span class="at">features =</span> <span class="fu">rownames</span>(sc_data))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>[…] and principal component analysis was performed.</p>
</blockquote>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(sc_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  S100P, SPINK1, KRT18, ADIRF, PSCA, KRT8, KRT19, MGST1, GDF15, KLF5 
       CD24, HPGD, DHRS2, HES1, VSIG2, SDC4, SMIM22, C19orf33, SLPI, S100A6 
       AGR2, SPINT2, TNFAIP2, GPRC5A, TSPAN6, S100A13, CLDN7, CD9, EFNA1, WFDC2 
Negative:  CD52, IGKC, LAPTM5, RGS1, S100A4, IGLC2, IL32, CD3D, CD74, IGHA1 
       IGHG1, CD69, LSP1, HSPA1A, VIM, LTB, CCL5, TRBC2, HLA-DPB1, COTL1 
       IGLC3, ALOX5AP, TRBC1, IGHG2, GPR183, RGS2, GZMA, TYROBP, NKG7, CD7 
PC_ 2 
Positive:  SPARC, IGFBP7, CALD1, COL1A2, COL3A1, COL6A2, COL4A1, IFITM3, MYL9, COL1A1 
       FN1, TPM2, CAVIN3, THY1, BGN, FSTL1, COX7A1, TIMP1, COL6A1, DCN 
       C1S, TAGLN, C1R, RGS5, COL5A2, LGALS1, SPARCL1, MFGE8, CTGF, COL6A3 
Negative:  HPGD, GATA3, CD24, S100P, RAB11FIP1, CD3D, SPINK1, PSCA, DHRS2, KRT8 
       KRT19, KLF5, CD52, UPK2, TRBC2, TMEM97, GDPD3, GDF15, TMPRSS2, VSIG2 
       UPK1A, SDC4, KRT20, SMIM22, SPTSSB, MGST1, CNGA1, WFDC2, CXADR, UPK3A 
PC_ 3 
Positive:  CALD1, SPARC, COL3A1, COL6A2, COL1A2, IGFBP7, IL32, COL4A1, COL1A1, RGS5 
       THY1, MYL9, TPM2, SELENOM, CAVIN3, COL6A1, FSTL1, SOD3, MFGE8, BGN 
       COL5A2, COX7A1, MGP, COL6A3, NDUFA4L2, SPARCL1, PCOLCE, CTGF, C1R, LHFPL6 
Negative:  TYROBP, FCER1G, LYZ, MS4A6A, CD14, HLA-DRA, FTL, CTSS, CD68, C1QC 
       C1QA, C1QB, MS4A7, FCGR2A, SPI1, OLR1, LST1, MS4A4A, C5AR1, PSAP 
       HLA-DRB5, SERPINA1, APOC1, CTSB, CST3, SAT1, CD163, HLA-DPA1, BCL2A1, HLA-DMA 
PC_ 4 
Positive:  RAMP2, PCAT19, VWF, CLEC14A, HSPG2, EGFL7, RAMP3, CALCRL, AQP1, ECSCR 
       PCDH17, PODXL, FLT1, CD34, CDH5, VWA1, FAM167B, CLDN5, SLCO2A1, CXorf36 
       ENG, PECAM1, CYYR1, MMRN2, NOTCH4, SLC9A3R2, INSR, HYAL2, TM4SF1, ESM1 
Negative:  COL1A2, COL3A1, COL1A1, DCN, COL6A3, TPM2, C1S, PCOLCE, C1R, COL5A2 
       TAGLN, LUM, SOD3, COL6A1, RARRES2, BGN, ACTA2, MYL9, AEBP1, SERPINF1 
       COL6A2, CYGB, MXRA8, RGS5, MFGE8, NDUFA4L2, MEG3, PPP1R14A, SDC2, SERPING1 
PC_ 5 
Positive:  RNASE1, UPK2, GDPD3, UBE2C, TMEM97, CNGA1, UPK1A, RAB11FIP1, S100A13, UPK3A 
       BHMT, TNFAIP2, STMN1, CD24, TMPRSS2, PLPP5, SDC4, BIRC5, NQO1, SPRR3 
       EMX2, KRT20, MKI67, ERBB2, CCNB2, CDK1, HMGB2, PCLAF, MYCL, MESP1 
Negative:  KRT17, FABP4, SERPINB5, DKK1, PLA2G2A, CLU, CLCA4, CRTAC1, LAMB3, S100A2 
       AQP3, TSPAN1, IGFBP2, IGFBP5, ERRFI1, IRF6, ANXA1, BTBD16, ANXA10, FHL2 
       SHH, PERP, FABP5, PRAC1, CYP1B1, FOSL1, HSD17B2, SOX15, SFN, GPR87 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>             used    (Mb) gc trigger    (Mb)   max used  (Mb)
Ncells   13337713   712.4   20465354  1093.0   20465354  1093
Vcells 1682605518 12837.3 5535281110 42230.9 5765847106 43990</code></pre>
</div>
</div>
<p>We can visualize the PCA with <code>DimPlot</code>, coloring by different categorical variables. Here, we see the influence of both the sample (batch) and CD45 positive vs negative (which is correlated to batch but might also contribute to variance).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sc_data, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"CD45"</span>), <span class="at">reduction =</span> <span class="st">"pca"</span>) <span class="sc">&amp;</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  theme_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-21-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="intro_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
<section id="dimensionality-determination" class="level4">
<h4 class="anchored" data-anchor-id="dimensionality-determination">Dimensionality determination</h4>
<blockquote class="blockquote">
<p>Dimensionality of the dataset was then assessed using the JackStraw and ElbowPlot functions.</p>
</blockquote>
<p>From Seurat documentation, we see that JackStraw is quite computationally intensive:</p>
<blockquote class="blockquote">
<p>In Macosko et al, we implemented a resampling test inspired by the JackStraw procedure. While still available in Seurat (see previous vignette), this is a slow and computationally expensive procedure, and we is no longer routinely used in single cell analysis.</p>
</blockquote>
<p>This is not the most beautiful Elbow plot ever, but the idea is to find the “inflection point” of the plot (I’m estimating at about 10) and go a little bit further into the “stable” area, where adding more dimensions doesn’t add more infromation.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(sc_data, <span class="at">ndims =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-22-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="intro_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="480"></a></p>
</figure>
</div>
</div>
</div>
<p>For methods that require the dimensionality, we’ll set it at 15.</p>
</section>
</section>
<section id="doublet-detection" class="level3">
<h3 class="anchored" data-anchor-id="doublet-detection">Doublet detection</h3>
<p>It’s never a bad idea to add some doublet detection.</p>
<p>Doublets are UMIs that contained more than one cell and thefore have a mixed profile.</p>
<p>This benchmark shows some of the main methods (though new ones may emerge): https://www.sciencedirect.com/science/article/pii/S2405471220304592</p>
<p>We’ll use cxds from scdc because it’s fast and reasonably accurate.</p>
<p>However, it uses a <code>SingleCellExperiment</code> standard. Luckily, we can easily convert between <code>SeuratObject</code> and <code>SingleCellExperiment</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>cxds_scores <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">SplitObject</span>(sc_data, <span class="at">split.by =</span> <span class="st">"sample"</span>), <span class="cf">function</span>(samp_sc) {</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> samp_sc <span class="sc">|&gt;</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.SingleCellExperiment</span>() <span class="sc">|&gt;</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cxds</span>()</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce<span class="sc">$</span>cxds_score)</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>             used    (Mb) gc trigger    (Mb)   max used  (Mb)
Ncells   13355251   713.3   20465354  1093.0   20465354  1093
Vcells 1682217444 12834.4 5535281110 42230.9 5765847106 43990</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cxds_scores) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>cxds_scores <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, cxds_scores)[<span class="fu">Cells</span>(sc_data)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="clustering" class="level2">
<h2 class="anchored" data-anchor-id="clustering">Clustering</h2>
<blockquote class="blockquote">
<p>Clusters were calculated […]</p>
</blockquote>
<p>Before clustering, we need to establish a neighborhood graph and compute shared nearest neighbors (SNN). We’ll do this on 15 PCs:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sc_data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
</div>
<p>From this, we can use “FindClusters” for Louvain clustering, which uses a rather temperamental “resolution” parameter, very dependent on datasets usually.</p>
<p>The largest the “resolution”, the more clusters are obtained.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sc_data, <span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 44699
Number of edges: 1534079

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9465
Number of communities: 26
Elapsed time: 8 seconds</code></pre>
</div>
</div>
<p>UMAP and t-SNE are dimensionality reductions for visualization. They are very popular and we’ll use them widely. They minimize local distances, often making “neat” looking clusters.</p>
<p>Be wary of not using the UMAP projection itself as an input for analysis, as they contain lots of data distortions (which is normal of anything that is reducing a space of 3000 features to 2 dimensions)</p>
<blockquote class="blockquote">
<p>[…] and data dimensions were reduced using the t-SNE and UMAP methods.</p>
</blockquote>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:47:24 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:47:24 Read 44699 rows and found 15 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:47:24 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:47:24 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
12:47:27 Writing NN index file to temp file /tmp/RtmpiH8I36/file7a2b4fbb2ae2
12:47:27 Searching Annoy index using 1 thread, search_k = 3000
12:47:39 Annoy recall = 100%
12:47:40 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
12:47:42 Initializing from normalized Laplacian + noise (using RSpectra)
12:47:44 Commencing optimization for 200 epochs, with 1900728 positive edges
12:47:57 Optimization finished</code></pre>
</div>
</div>
<p>We can check the state of our Seurat object:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>sc_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
33538 features across 44699 samples within 1 assay 
Active assay: RNA (33538 features, 2000 variable features)
 21 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, data.1, data.2, data.3, data.4, data.5, data.6, data.7, data.8, data.9, data.10, scale.data
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<p>And plot the Seurat clusters… only to find we’re definitely clustering per sample.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sc_data, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"unintegrated_clusters"</span>, <span class="st">"sample"</span>), <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">&amp;</span> </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  theme_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-28-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="intro_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="integration" class="level2">
<h2 class="anchored" data-anchor-id="integration">Integration</h2>
<p>Batch effect is rampant in single-cell RNA-seq and there are lots of integration methods proposed to deal with this. They often return a dimensionality reduction (more precisely, latent space) that minimizes this batch effect.</p>
<p>There are 5 methods implemented in Seurat alone. We’ll use Harmony, because it’s fairly fast and effective.</p>
<p>See the reference tutorial: https://satijalab.org/seurat/articles/seurat5_integration</p>
<blockquote class="blockquote">
<p>Bulk cells were merged for the HLA analysis, while CD8 T cells only were merged for the CD8 T cell analysis. CD8 T cells were identified using the FindAllMarkers and the VlnPlot functions.</p>
</blockquote>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(sc_data, </span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">new.reduction =</span> <span class="st">"harmony"</span>,</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in harmony::HarmonyMatrix(data_mat = Embeddings(object = orig), :
HarmonyMatrix is deprecated and will be removed in the future from the API in
the future</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameters do_pca and npcs are deprecated. They will be ignored for this function call and please remove parameters do_pca and npcs and pass to harmony cell_embeddings directly.
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter tau is deprecated. It will be ignored for this function call and please remove parameter tau in future function calls. Advanced users can set value of parameter tau by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter block.size is deprecated. It will be ignored for this function call and please remove parameter block.size in future function calls. Advanced users can set value of parameter block.size by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter max.iter.harmony is replaced with parameter max_iter. It will be ignored for this function call and please use parameter max_iter in future function calls.
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter max.iter.cluster is deprecated. It will be ignored for this function call and please remove parameter max.iter.cluster in future function calls. Advanced users can set value of parameter max.iter.cluster by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter epsilon.cluster is deprecated. It will be ignored for this function call and please remove parameter epsilon.cluster in future function calls. Advanced users can set value of parameter epsilon.cluster by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter epsilon.harmony is deprecated. It will be ignored for this function call and please remove parameter epsilon.harmony in future function calls. If users want to control if harmony would stop early or not, use parameter early_stop. Advanced users can set value of parameter epsilon.harmony by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>             used    (Mb) gc trigger    (Mb)   max used  (Mb)
Ncells   13450619   718.4   20465354  1093.0   20465354  1093
Vcells 1691091272 12902.1 5535281110 42230.9 5765847106 43990</code></pre>
</div>
</div>
<p>We’ll need re-run clustering algorithm in the Harmony latent space (here, using 30 dimensions)</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sc_data, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sc_data, <span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">cluster.name =</span> <span class="st">"harmony_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 44699
Number of edges: 1710955

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9238
Number of communities: 18
Elapsed time: 10 seconds</code></pre>
</div>
</div>
<p>We can also run a UMAP in this latent space:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction.name =</span> <span class="st">"umap_harmony"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>12:49:29 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:49:29 Read 44699 rows and found 30 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:49:29 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:49:29 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
12:49:31 Writing NN index file to temp file /tmp/RtmpiH8I36/file7a2b2a001d88
12:49:31 Searching Annoy index using 1 thread, search_k = 3000
12:49:44 Annoy recall = 100%
12:49:44 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
12:49:46 Initializing from normalized Laplacian + noise (using RSpectra)
12:49:48 Commencing optimization for 200 epochs, with 2061946 positive edges
12:50:03 Optimization finished</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sc_data, </span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"unintegrated_clusters"</span>, <span class="st">"harmony_clusters"</span>, <span class="st">"sample"</span>), </span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">alpha =</span> <span class="fl">0.2</span>,</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">&amp;</span> </span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>  theme_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-32-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="intro_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<section id="check-biases" class="level3">
<h3 class="anchored" data-anchor-id="check-biases">Check biases</h3>
<p>We can check back on our features that could be markers of bias in a cluster, and check back in with our doublet scores:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>cxds_scores[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>AAACCTGAGGCCCTCA-1_1 AAACCTGAGTTCGCGC-1_1 AAACCTGCACTGTTAG-1_1 
           266320.09             29805.91            219085.93 
AAACCTGCAGACGCTC-1_1 AAACCTGGTACAGCAG-1_1 
           100458.58            193619.85 </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(sc_data, </span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, </span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"percent.mt"</span>,<span class="st">"cxds_scores"</span>), </span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">&amp;</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">&amp;</span></span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>  theme_clean </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 569</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 123</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.469</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 6840</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-34-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="intro_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>We can see, for example, that cluster 17 seems to be enriched in doublets.</p>
<p>To compare the co-occurence of two categorical variables, we can do a heatmap:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sc_data<span class="sc">$</span>harmony_clusters, sc_data<span class="sc">$</span>sample) <span class="sc">|&gt;</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pheatmap</span>(<span class="at">scale =</span> <span class="st">"row"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-35-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="intro_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>From this, we can exclude cluster 17:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">subset</span>(sc_data, <span class="at">subset =</span> harmony_clusters <span class="sc">!=</span> <span class="dv">17</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="annotation" class="level2">
<h2 class="anchored" data-anchor-id="annotation">Annotation</h2>
<section id="differential-expression" class="level3">
<h3 class="anchored" data-anchor-id="differential-expression">Differential Expression</h3>
<p>Differential expression analysis will by default run on the “identity” feature, which we make sure to set as “harmony_clusters” at this point:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(sc_data) <span class="ot">&lt;-</span> sc_data<span class="sc">$</span>harmony_clusters</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>AAACCTGAGGCCCTCA-1_1 AAACCTGAGTTCGCGC-1_1 AAACCTGCACTGTTAG-1_1 
                   2                    6                    1 
AAACCTGCAGACGCTC-1_1 AAACCTGGTACAGCAG-1_1 
                   3                    1 
Levels: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16</code></pre>
</div>
</div>
<p>For differential analysis to use all layers at the same time, we need to join them before finding markers (each clusters vs all others):</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(sc_data)</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(sc_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 9</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 11</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 12</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 13</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 15</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 16</code></pre>
</div>
</div>
<p>For annotation, we can now use this table:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>markers <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene
IL7R     0   2.929902 0.634 0.158         0       0 IL7R
CD3D     0   1.512433 0.750 0.281         0       0 CD3D
IL32     0   1.225477 0.865 0.404         0       0 IL32
TRAC     0   1.525893 0.722 0.273         0       0 TRAC
LTB      0   2.055308 0.713 0.273         0       0  LTB
CD2      0   1.805078 0.660 0.230         0       0  CD2</code></pre>
</div>
</div>
<p>And visualize either top markers or any chosen features in many different ways:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>top_markers <span class="ot">&lt;-</span> markers <span class="sc">|&gt;</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">|&gt;</span></span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(p_val_adj) <span class="sc">|&gt;</span></span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>top5 <span class="ot">&lt;-</span> <span class="fu">unique</span>(top_markers<span class="sc">$</span>gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="heatmap" class="level4">
<h4 class="anchored" data-anchor-id="heatmap">Heatmap</h4>
<p>The heatmap takes a while to run, but provides a comprehensive view of a marker expression over the cells:</p>
<div class="cell" data-fig.heigth="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(sc_data, <span class="at">features =</span> top5, <span class="at">raster =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-41-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="intro_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="dot-plot" class="level4">
<h4 class="anchored" data-anchor-id="dot-plot">Dot plot</h4>
<p>The dotplot summarizes the information into the heatmap and gives the same importance to all clusters, independent of size:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(sc_data, <span class="at">features =</span> top5, <span class="at">cluster.idents =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> theme_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-42-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="intro_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="1440"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="featureplot" class="level4">
<h4 class="anchored" data-anchor-id="featureplot">FeaturePlot</h4>
<p><code>FeaturePlot</code> is the same as <code>DimPlot</code>, but using one or many continuous values instead:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(sc_data, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"MS4A1"</span>,</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"CD14"</span>, <span class="st">"PECAM1"</span>,</span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"KRT20"</span>, <span class="st">"KRT5"</span>)) <span class="sc">&amp;</span></span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-43-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="intro_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="reference-based" class="level3">
<h3 class="anchored" data-anchor-id="reference-based">Reference-based</h3>
<p>There are an infinity of methods for cell type annotation. I’ve found that using a good reference is more important than the method for annotation.</p>
<p>We’ll show reference-based annotation using SingleR, mostly because it’s easy to use but also well performing.</p>
<section id="built-in-references" class="level4">
<h4 class="anchored" data-anchor-id="built-in-references">“Built-in” references</h4>
<p>Every automatic annotation method uses a reference. <code>celldex</code> provides an easy way to access some annotation sources, notably the Human Primary Cell Atlas, which we will use in this example.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(celldex)</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">surveyReferences</span>() <span class="sc">|&gt;</span>  <span class="fu">as_tibble</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 10
  name         version path  title description taxonomy_id genome samples labels
  &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;       &lt;list&gt;      &lt;list&gt;   &lt;int&gt; &lt;list&gt;
1 dice         2024-0… &lt;NA&gt;  Huma… Human bulk… &lt;list [1]&gt;  &lt;list&gt;    1561 &lt;list&gt;
2 blueprint_e… 2024-0… &lt;NA&gt;  Huma… Human bulk… &lt;list [1]&gt;  &lt;list&gt;     259 &lt;list&gt;
3 immgen       2024-0… &lt;NA&gt;  Mous… Mouse micr… &lt;list [1]&gt;  &lt;list&gt;     830 &lt;list&gt;
4 mouse_rnaseq 2024-0… &lt;NA&gt;  Bulk… Bulk RNA-s… &lt;list [1]&gt;  &lt;list&gt;     358 &lt;list&gt;
5 hpca         2024-0… &lt;NA&gt;  Micr… Microarray… &lt;list [1]&gt;  &lt;list&gt;     713 &lt;list&gt;
6 novershtern… 2024-0… &lt;NA&gt;  Bulk… Bulk micro… &lt;list [1]&gt;  &lt;list&gt;     211 &lt;list&gt;
7 monaco_immu… 2024-0… &lt;NA&gt;  Huma… Human bulk… &lt;list [1]&gt;  &lt;list&gt;     114 &lt;list&gt;
# ℹ 1 more variable: sources &lt;list&gt;</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>hpca_se <span class="ot">&lt;-</span> celldex<span class="sc">::</span><span class="fu">HumanPrimaryCellAtlasData</span>()</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>hpca_se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 19363 713 
metadata(0):
assays(1): logcounts
rownames(19363): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
rowData names(0):
colnames(713): GSM112490 GSM112491 ... GSM92233 GSM92234
colData names(3): label.main label.fine label.ont</code></pre>
</div>
</div>
<p>It’s given as a <code>SummarizedExperiment</code> of purified bulk tumors.=, annotated as follows:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(hpca_se<span class="sc">$</span>label.main)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "DC"                   "Smooth_muscle_cells"  "Epithelial_cells"    
 [4] "B_cell"               "Neutrophils"          "T_cells"             
 [7] "Monocyte"             "Erythroblast"         "BM &amp; Prog."          
[10] "Endothelial_cells"    "Gametocytes"          "Neurons"             
[13] "Keratinocytes"        "HSC_-G-CSF"           "Macrophage"          
[16] "NK_cell"              "Embryonic_stem_cells" "Tissue_stem_cells"   
[19] "Chondrocytes"         "Osteoblasts"          "BM"                  
[22] "Platelets"            "Fibroblasts"          "iPS_cells"           
[25] "Hepatocytes"          "MSC"                  "Neuroepithelial_cell"
[28] "Astrocyte"            "HSC_CD34+"            "CMP"                 
[31] "GMP"                  "MEP"                  "Myelocyte"           
[34] "Pre-B_cell_CD34-"     "Pro-B_cell_CD34+"     "Pro-Myelocyte"       </code></pre>
</div>
</div>
<p>There’s also a finer annotation, which we won’t use:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(hpca_se<span class="sc">$</span>label.fine) <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "DC:monocyte-derived:immature"        "DC:monocyte-derived:Galectin-1"     
 [3] "DC:monocyte-derived:LPS"             "DC:monocyte-derived"                
 [5] "Smooth_muscle_cells:bronchial:vit_D" "Smooth_muscle_cells:bronchial"      
 [7] "Epithelial_cells:bronchial"          "B_cell"                             
 [9] "Neutrophil"                          "T_cell:CD8+_Central_memory"         
[11] "T_cell:CD8+"                         "T_cell:CD4+"                        
[13] "T_cell:CD8+_effector_memory_RA"      "T_cell:CD8+_effector_memory"        
[15] "T_cell:CD8+_naive"                   "Monocyte"                           
[17] "Erythroblast"                        "BM"                                 
[19] "DC:monocyte-derived:rosiglitazone"   "DC:monocyte-derived:AM580"          </code></pre>
</div>
</div>
<p>SingleR is really simple to use, we just need to provide an expression matrix and reference as <code>SummarizedExperiment</code>:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleR)</span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>pred_hpca <span class="ot">&lt;-</span> <span class="fu">SingleR</span>(<span class="at">test =</span> <span class="fu">GetAssayData</span>(sc_data), <span class="at">ref =</span> hpca_se, <span class="at">assay.type.test=</span><span class="dv">1</span>,</span>
<span id="cb161-3"><a href="#cb161-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> hpca_se<span class="sc">$</span>label.main)</span>
<span id="cb161-4"><a href="#cb161-4" aria-hidden="true" tabindex="-1"></a>pred_hpca</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 44535 rows and 4 columns
                                                scores              labels
                                              &lt;matrix&gt;         &lt;character&gt;
AAACCTGAGGCCCTCA-1_1   0.2029037:0.213122:0.201966:...    Epithelial_cells
AAACCTGAGTTCGCGC-1_1   0.0681688:0.198615:0.175877:...              B_cell
AAACCTGCACTGTTAG-1_1   0.0876658:0.201213:0.204570:...             T_cells
AAACCTGCAGACGCTC-1_1   0.1877214:0.195236:0.193748:...    Epithelial_cells
AAACCTGGTACAGCAG-1_1   0.0911602:0.192878:0.188423:...             T_cells
...                                                ...                 ...
TTTGTCAGTAAGTAGT-1_10 0.1291041:0.126240:0.0979679:...   Endothelial_cells
TTTGTCAGTCCTCCAT-1_10 0.2087580:0.136874:0.1276952:... Smooth_muscle_cells
TTTGTCAGTCTAGTCA-1_10 0.1394180:0.249968:0.2645048:...            Monocyte
TTTGTCAGTGCTGTAT-1_10 0.2323622:0.210428:0.1944528:...   Endothelial_cells
TTTGTCATCTAGCACA-1_10 0.0968535:0.202014:0.1887533:...             NK_cell
                      delta.next       pruned.labels
                       &lt;numeric&gt;         &lt;character&gt;
AAACCTGAGGCCCTCA-1_1   0.0539724    Epithelial_cells
AAACCTGAGTTCGCGC-1_1   0.0693563              B_cell
AAACCTGCACTGTTAG-1_1   0.1751957             T_cells
AAACCTGCAGACGCTC-1_1   0.5717106    Epithelial_cells
AAACCTGGTACAGCAG-1_1   0.0322466             T_cells
...                          ...                 ...
TTTGTCAGTAAGTAGT-1_10 0.05911158   Endothelial_cells
TTTGTCAGTCCTCCAT-1_10 0.00671502 Smooth_muscle_cells
TTTGTCAGTCTAGTCA-1_10 0.02419472            Monocyte
TTTGTCAGTGCTGTAT-1_10 0.06293215   Endothelial_cells
TTTGTCATCTAGCACA-1_10 0.09986082             NK_cell</code></pre>
</div>
</div>
<p>Once we have the predictions, I add it to the Seurat object:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">@</span>tools[[<span class="st">"SingleR"</span>]] <span class="ot">&lt;-</span> pred_hpca <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>singleR_pred <span class="ot">&lt;-</span> pred_hpca<span class="sc">$</span>pruned.labels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="harmonizing-automatic-predictions" class="level5">
<h5 class="anchored" data-anchor-id="harmonizing-automatic-predictions">Harmonizing automatic predictions</h5>
<p>We can check the prediction results and “harmonize” them by cluster, i.e.&nbsp;take the cluster-level mode of the annotation:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>harmony_singler <span class="ot">&lt;-</span> <span class="fu">table</span>(sc_data<span class="sc">$</span>singleR_pred, <span class="fu">factor</span>(sc_data<span class="sc">$</span>harmony_clusters))</span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(harmony_singler, <span class="at">scale =</span> <span class="st">"column"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-50-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="intro_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>harmony_map1 <span class="ot">&lt;-</span> <span class="fu">apply</span>(harmony_singler, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">names</span>(<span class="fu">which.max</span>(x)))</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a>harmony_map1 <span class="ot">&lt;-</span> <span class="fu">structure</span>(<span class="fu">names</span>(harmony_map1), <span class="at">names =</span> harmony_map1)</span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a>harmony_map1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>          T_cells           T_cells  Epithelial_cells  Epithelial_cells 
              "0"               "1"               "2"               "3" 
         Monocyte        Macrophage            B_cell Tissue_stem_cells 
              "4"               "5"               "6"               "7" 
          T_cells            B_cell  Epithelial_cells  Epithelial_cells 
              "8"               "9"              "10"              "11" 
Endothelial_cells          Monocyte               CMP  Pre-B_cell_CD34- 
             "12"              "13"              "14"              "15" 
               DC 
             "16" </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>singleR_clean <span class="ot">&lt;-</span> <span class="fu">fct_recode</span>(sc_data<span class="sc">$</span>harmony_clusters, <span class="sc">!!!</span> harmony_map1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>If we plot, we can see the difference between the “pred” and “clean”:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sc_data, </span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"harmony_clusters"</span>, <span class="st">"singleR_pred"</span>, </span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"singleR_clean"</span>, <span class="st">"sample"</span>), </span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,</span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb168-6"><a href="#cb168-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">&amp;</span> </span>
<span id="cb168-7"><a href="#cb168-7" aria-hidden="true" tabindex="-1"></a>  theme_clean <span class="sc">&amp;</span></span>
<span id="cb168-8"><a href="#cb168-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_igv</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 67 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-53-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="intro_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="single-cell" class="level4">
<h4 class="anchored" data-anchor-id="single-cell">Single-cell</h4>
<p>Most methods will work best with a relevant single-cell reference.</p>
<p>Here is a not-so-recent but independent benchmark: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1795-z</p>
<p>It can be interesting to try to use this tutorial to annotate a query dataset based on a reference atlas: https://satijalab.org/seurat/articles/integration_mapping</p>
</section>
</section>
</section>
<section id="extras" class="level2">
<h2 class="anchored" data-anchor-id="extras">Extras</h2>
<section id="sub-cluster" class="level3">
<h3 class="anchored" data-anchor-id="sub-cluster">Sub-cluster</h3>
<p>In their manuscript, Salomé et al selected the T cells and subclustered them. This is a common strategy when there is a main population of interest. In fact, most of their analysis concentrated on the CD8 T cells.</p>
<p>At this point, we could take our annotated data, subset the T cells and re-run the pipeline to find new relevant subgroups.</p>
</section>
<section id="trajectory-analysis" class="level3">
<h3 class="anchored" data-anchor-id="trajectory-analysis">Trajectory analysis</h3>
<p>Trajectory analysis is used to reconstruct and visualize the dynamic processes of cellular development, differentiation, and state transitions. It infers the order of cells in a particular trajectory, allows us to calculate continuous “pseudotimes”.</p>
<p>It can be used to map out potential lineage relationships, identify intermediate cell states, and elucidate the pathways underlying tissue development, immune responses, or disease progression, for example.</p>
<p>In Figure 4, Salomé et al used this to show a subset of differentiated PD-1+ TRM CD8+ T Cells that retain independent function.</p>
<p><a href="images/Screenshot from 2025-02-13 18-42-25.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="images/Screenshot from 2025-02-13 18-42-25.png" class="img-fluid"></a></p>
<p>A not so new but very comprehensive benchmark of trajectory inference by Saelens et al is a great jumping point: <a href="https://www.nature.com/articles/s41587-019-0071-9" class="uri">https://www.nature.com/articles/s41587-019-0071-9</a></p>
<p>Accompanied by the superb dynverse wrappers, which provide all methods in the benchmark, as well as guidance on what to choose: <a href="https://dynverse.org/" class="uri">https://dynverse.org/</a></p>
</section>
<section id="cell-cycle" class="level3">
<h3 class="anchored" data-anchor-id="cell-cycle">Cell cycle</h3>
<p>There are many methods to infer cell cycle stage and score cell cycle in Seurat, but it also has a built in one that works quite decently:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(sc_data, </span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">s.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>s.genes, </span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">g2m.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>g2m.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(sc_data, </span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>),</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,</span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb171-5"><a href="#cb171-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">&amp;</span> </span>
<span id="cb171-6"><a href="#cb171-6" aria-hidden="true" tabindex="-1"></a>  theme_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><a href="intro_files/figure-html/unnamed-chunk-55-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="intro_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
<p>We can see very little cycling except in one “epithelial” (tumor) area we previously annotated.</p>
</section>
<section id="rna-velocity" class="level3">
<h3 class="anchored" data-anchor-id="rna-velocity">RNA velocity</h3>
<p>RNA velocity is a computation method that attempts to predict the future transcriptional state of individual cells. It distinguishes between unspliced (pre-mRNA) and spliced (mature mRNA) transcripts within each cell, and therefore necessitates a special input in which these reads have been separated. This differentiation allows the estimation of the rate at which genes are being transcribed and processed, providing insights into the “direction” and “speed” of cellular state changes.</p>
<p>The primary goal of RNA velocity analysis is to elucidate the trajectories of cellular differentiation, activation, and other dynamic processes, thereby offering a deeper understanding of cellular development and lineage relationships. By forecasting how cells are likely to evolve, it can help in identifying transient cell states, and uncovering mechanisms underlying tissue development, regeneration, and disease progression.</p>
<p>It can be used together with traditional trajectory analysis.</p>
<p>Most RNA velocity tools work in Python, the two most well-known being <code>velocyto</code> and <code>scVelo</code>. Both have tutorials showing how to use reticulate to run their workflows in R.</p>
</section>
<section id="pathway-enrichment" class="level3">
<h3 class="anchored" data-anchor-id="pathway-enrichment">Pathway enrichment</h3>
<p>Pathway enrichment aims to identify biological pathways, processes, or functional categories that are significantly overrepresented within a set of genes of interest, such as those differentially expressed between distinct cell populations or states.</p>
<p>This analysis typically involves mapping the selected genes to curated pathway databases like Gene Ontology, KEGG, or Reactome and applying statistical methods to assess the significance of their enrichment.</p>
<p>There are many methods to do this, but a easy to use is decoupleR, in which AUCell is implemented, which is a well-performing scoring method: <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/decoupleR/inst/doc/decoupleR.html" class="uri">https://www.bioconductor.org/packages/release/bioc/vignettes/decoupleR/inst/doc/decoupleR.html</a></p>
</section>
<section id="gene-regulatory-network" class="level3">
<h3 class="anchored" data-anchor-id="gene-regulatory-network">Gene regulatory network</h3>
<p>Gene regulatory network (GRN) analysis seeks to identify key transcription factors, regulatory motifs, and signaling pathways that drive cell states, differentiation, and responses to environmental stimuli. The ultimate goal of GRN analysis is to map out the regulatory architecture that underpins cellular heterogeneity, enabling \ to pinpoint master regulators and understand the molecular mechanisms underlying various biological processes and disease states.</p>
<p>From an algorithmic point of view, this involves inferring relationships such as gene co-expression, regulatory influences, and causal interactions.</p>
<p>SCENIC is the most popular method to do this, with the Python implementation being much more performant (<a href="https://github.com/aertslab/pySCENIC" class="uri">https://github.com/aertslab/pySCENIC</a>). The protocol is probably the best way to run this type of analysis currently: <a href="https://github.com/aertslab/SCENICprotocol" class="uri">https://github.com/aertslab/SCENICprotocol</a></p>
</section>
<section id="cnv-analysis" class="level3">
<h3 class="anchored" data-anchor-id="cnv-analysis">CNV analysis</h3>
<p>Copy Number Variation (CNV) analysis is a computational approach used to infer genomic copy number alterations from transcriptomic data. It leverages the variability in gene expression levels across the genome to detect gains or losses of genomic regions.</p>
<p>This method is particularly valuable in studies of cancer and other genetic disorders, where CNVs play a crucial role in disease progression and heterogeneity.</p>
<p>The most well-known tool used for this analysis is inferCNV, which can be extremely slow and doesn’t perform very well with cells with low-read count. This is why we implemented FastCNV, available to install for R on Github: <a href="https://github.com/must-bioinfo/fastCNVdata" class="uri">https://github.com/must-bioinfo/fastCNVdata</a></p>
</section>
<section id="ligand-receptor-analysis" class="level3">
<h3 class="anchored" data-anchor-id="ligand-receptor-analysis">Ligand-receptor analysis</h3>
<p>Ligand-receptor analysis is a computational strategy to infer cell-cell communication by identifying interactions between signaling molecules (ligands) and their corresponding receptors expressed on different cell types or states.</p>
<p>This analysis involves systematically pairing ligands expressed by one or more cell populations with receptors expressed by neighboring or interacting cells, thereby predicting potential signaling pathways that facilitate intercellular communication.</p>
<p>Well known tools are CellPhoneDB and CellChat, but LIANA+ provides wrappers and an all-in-one framework for this kind of analysis: <a href="https://liana-py.readthedocs.io/en/latest/" class="uri">https://liana-py.readthedocs.io/en/latest/</a></p>
<p>Notably, this analysis can be much more powerful in spatially-resolved data.</p>
</section>
<section id="object-conversion" class="level3">
<h3 class="anchored" data-anchor-id="object-conversion">Object conversion</h3>
<p>Format conversion is often necessary when working with single-cell RNA-seq to use tools available for different objects.</p>
<p>This can be easy between SingleCellExperiment and SeuratObject, as we used it in this tutorial with <code>as.SingleCellExperiment</code> and <code>as.Seurat</code>.</p>
<p>However, these functions sometimes don’t work and the Seurat solution for h5 objects, SeuratDisk, fails its own tutorial. <code>sceasy</code> (https://github.com/cellgeni/sceasy) was made to resolve these conversion problems.</p>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb172" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> R and **Seurat** to analyse single cell RNA-seq</span></span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> Case study - NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer</span></span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Clarice Groeneveld</span></span>
<span id="cb172-5"><a href="#cb172-5" aria-hidden="true" tabindex="-1"></a><span class="an">editor:</span><span class="co"> </span></span>
<span id="cb172-6"><a href="#cb172-6" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb172-7"><a href="#cb172-7" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb172-8"><a href="#cb172-8" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb172-9"><a href="#cb172-9" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb172-10"><a href="#cb172-10" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb172-11"><a href="#cb172-11" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-depth: 4</span></span>
<span id="cb172-12"><a href="#cb172-12" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: Contents</span></span>
<span id="cb172-13"><a href="#cb172-13" aria-hidden="true" tabindex="-1"></a><span class="co">    lightbox: true</span></span>
<span id="cb172-14"><a href="#cb172-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: true</span></span>
<span id="cb172-15"><a href="#cb172-15" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb172-16"><a href="#cb172-16" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb172-17"><a href="#cb172-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-18"><a href="#cb172-18" aria-hidden="true" tabindex="-1"></a>Download this notebook: https://raw.githubusercontent.com/csgroen/teaching/refs/heads/main/20250214_Tutorial_DU_Seurat.qmd</span>
<span id="cb172-19"><a href="#cb172-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-20"><a href="#cb172-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pre-requisites</span></span>
<span id="cb172-21"><a href="#cb172-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-22"><a href="#cb172-22" aria-hidden="true" tabindex="-1"></a>For the best understanding of this tutorial, please make sure you have at least an intermediate understanding of R. Here are some recommended resources to learn the basics and necessary intermediate concepts:</span>
<span id="cb172-23"><a href="#cb172-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-24"><a href="#cb172-24" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>swirl: https://swirlstats.com/students.html</span>
<span id="cb172-25"><a href="#cb172-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-26"><a href="#cb172-26" aria-hidden="true" tabindex="-1"></a>Specifically courses *R Programming*, *Getting and Cleaning Data* and *Exploratory Data Analysis* may provide a good jump start.</span>
<span id="cb172-27"><a href="#cb172-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-28"><a href="#cb172-28" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Hands On Programming with R: https://rstudio-education.github.io/hopr/</span>
<span id="cb172-29"><a href="#cb172-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-30"><a href="#cb172-30" aria-hidden="true" tabindex="-1"></a>The very basics of R programming as also explained in this thorough book from the Posit team (that builds RStudio).</span>
<span id="cb172-31"><a href="#cb172-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-32"><a href="#cb172-32" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>R for Data Science: https://r4ds.hadley.nz/</span>
<span id="cb172-33"><a href="#cb172-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-34"><a href="#cb172-34" aria-hidden="true" tabindex="-1"></a>This goes further than we need, but a base for most of the programming questions a data analyst working in R could have to start with.</span>
<span id="cb172-35"><a href="#cb172-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-36"><a href="#cb172-36" aria-hidden="true" tabindex="-1"></a><span class="fu">## Resources</span></span>
<span id="cb172-37"><a href="#cb172-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-38"><a href="#cb172-38" aria-hidden="true" tabindex="-1"></a>Seurat has many very useful tutorials for all of its features. Elements of this case study were taken from the following:</span>
<span id="cb172-39"><a href="#cb172-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-40"><a href="#cb172-40" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Seurat - Guided Clustering Tutorial https://satijalab.org/seurat/archive/v3.0/pbmc3k_tutorial.html</span>
<span id="cb172-41"><a href="#cb172-41" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Introduction to scRNA-seq integration: https://satijalab.org/seurat/articles/integration_introduction</span>
<span id="cb172-42"><a href="#cb172-42" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Mapping and annotating query datasets: https://satijalab.org/seurat/articles/integration_mapping</span>
<span id="cb172-43"><a href="#cb172-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-44"><a href="#cb172-44" aria-hidden="true" tabindex="-1"></a><span class="fu">## Install packages</span></span>
<span id="cb172-45"><a href="#cb172-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-46"><a href="#cb172-46" aria-hidden="true" tabindex="-1"></a>To ensure we have all installed libraries, we'll use R package "pacman", which allows us to check if packages are not installed and load them. This is not a replacement for more comprehensive version control (e.g. using <span class="in">`renv`</span>), but it's simple to install and very lightweight.</span>
<span id="cb172-47"><a href="#cb172-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-50"><a href="#cb172-50" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-51"><a href="#cb172-51" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span> <span class="st">"pacman"</span> <span class="sc">%in%</span> <span class="fu">installed.packages</span>()) <span class="fu">install.packages</span>(<span class="st">"pacman"</span>)</span>
<span id="cb172-52"><a href="#cb172-52" aria-hidden="true" tabindex="-1"></a>pacman<span class="sc">::</span><span class="fu">p_load</span>(<span class="st">"Seurat"</span>, <span class="st">"tidyverse"</span>, <span class="st">"R.utils"</span>, <span class="st">"ggpubr"</span>, <span class="st">"patchwork"</span>, </span>
<span id="cb172-53"><a href="#cb172-53" aria-hidden="true" tabindex="-1"></a>               <span class="st">"celldex"</span>, <span class="st">"SingleR"</span>, <span class="st">"scRNAseq"</span>, <span class="st">"pheatmap"</span>,</span>
<span id="cb172-54"><a href="#cb172-54" aria-hidden="true" tabindex="-1"></a>               <span class="st">"ggsci"</span>, <span class="st">"scds"</span>)</span>
<span id="cb172-55"><a href="#cb172-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-56"><a href="#cb172-56" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-57"><a href="#cb172-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-58"><a href="#cb172-58" aria-hidden="true" tabindex="-1"></a>I'm personally not a big fan of the standard look of Seurat plots, so I'll apply this ggplot2 theme to (almost) all Seurat plots we generate:</span>
<span id="cb172-59"><a href="#cb172-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-62"><a href="#cb172-62" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-63"><a href="#cb172-63" aria-hidden="true" tabindex="-1"></a>theme_clean <span class="ot">&lt;-</span></span>
<span id="cb172-64"><a href="#cb172-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>() <span class="sc">+</span></span>
<span id="cb172-65"><a href="#cb172-65" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb172-66"><a href="#cb172-66" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb172-67"><a href="#cb172-67" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">1</span>))</span>
<span id="cb172-68"><a href="#cb172-68" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-69"><a href="#cb172-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-70"><a href="#cb172-70" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Note: Seurat plots are ggplot-based and therefore can be "added to" in a grammar-of-graphics manner. Plots for multiple features are patchworks, which use a slightly different syntax to add to (this is why sometimes you'll see </span><span class="in">`&amp;`</span><span class="at"> instead of </span><span class="in">`+`</span><span class="at"> being used).</span></span>
<span id="cb172-71"><a href="#cb172-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-72"><a href="#cb172-72" aria-hidden="true" tabindex="-1"></a><span class="fu">## Download data</span></span>
<span id="cb172-73"><a href="#cb172-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-74"><a href="#cb172-74" aria-hidden="true" tabindex="-1"></a>Our dataset comes from Salomé et al (2022) <span class="sc">\[</span>https://www.sciencedirect.com/science/article/pii/S1535610822003695<span class="sc">\]</span>, "NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer". This single-cell data is a bit particular: they have FACS-sorted their cells into CD45+ and CD45-, notably to enrich their dataset in hematopoetic cells.</span>
<span id="cb172-75"><a href="#cb172-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-76"><a href="#cb172-76" aria-hidden="true" tabindex="-1"></a>We'll download the data from the link provided in the article:</span>
<span id="cb172-77"><a href="#cb172-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-80"><a href="#cb172-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-81"><a href="#cb172-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb172-82"><a href="#cb172-82" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>)</span>
<span id="cb172-83"><a href="#cb172-83" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://prod-dcd-datasets-cache-zipfiles.s3.eu-west-1.amazonaws.com/7yb7s9769c-1.zip"</span>, <span class="st">'data/Salome_2022.zip'</span>)</span>
<span id="cb172-84"><a href="#cb172-84" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(<span class="st">'data/Salome_2022.zip'</span>, <span class="at">exdir =</span> <span class="st">"data"</span>)</span>
<span id="cb172-85"><a href="#cb172-85" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-86"><a href="#cb172-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-87"><a href="#cb172-87" aria-hidden="true" tabindex="-1"></a><span class="fu">## Read 10X data</span></span>
<span id="cb172-88"><a href="#cb172-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-91"><a href="#cb172-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-92"><a href="#cb172-92" aria-hidden="true" tabindex="-1"></a>ref_dir <span class="ot">&lt;-</span> <span class="st">"data/NKG2A and HLA-E define an alternative immune checkpoint axis in bladder cancer/"</span></span>
<span id="cb172-93"><a href="#cb172-93" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(ref_dir)</span>
<span id="cb172-94"><a href="#cb172-94" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-95"><a href="#cb172-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-96"><a href="#cb172-96" aria-hidden="true" tabindex="-1"></a>We can see the data is distributed as one folder per dataset, and includes the bulk data, which we're not going to use. For this reason, I will filter out the "bulk" folders and to get my vector of <span class="in">`sc_dts`</span> (single-cell datasets).</span>
<span id="cb172-97"><a href="#cb172-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-100"><a href="#cb172-100" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-101"><a href="#cb172-101" aria-hidden="true" tabindex="-1"></a>sc_dts <span class="ot">&lt;-</span> ref_dir <span class="sc">|&gt;</span> <span class="fu">list.files</span>() <span class="sc">|&gt;</span> <span class="fu">str_subset</span>(<span class="st">"bulk"</span>, <span class="at">negate =</span> <span class="cn">TRUE</span>)</span>
<span id="cb172-102"><a href="#cb172-102" aria-hidden="true" tabindex="-1"></a>sc_dts</span>
<span id="cb172-103"><a href="#cb172-103" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-104"><a href="#cb172-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-105"><a href="#cb172-105" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; **Syntax note:** </span><span class="in">`ref_dir |&gt; list.files() |&gt; str_subset("bulk", negate = TRUE)`</span><span class="at"> is equivalent to </span><span class="in">`str_subset(list.files(ref_dir), "bulk", negate = TRUE)`</span><span class="at">. The </span><span class="in">`|&gt;`</span><span class="at"> was first introduced as </span><span class="in">`%&gt;%`</span><span class="at"> from R package magrittr and officially incorporated into base R</span><span class="sc">\&gt;</span><span class="at">=4.0. It's purpose is to increase code legibility through "chaining" commands instead of "nesting" them. If you would like to learn more, you can check the help file </span><span class="in">`? |&gt;`</span></span>
<span id="cb172-106"><a href="#cb172-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-107"><a href="#cb172-107" aria-hidden="true" tabindex="-1"></a>We can check inside each folder:</span>
<span id="cb172-108"><a href="#cb172-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-111"><a href="#cb172-111" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-112"><a href="#cb172-112" aria-hidden="true" tabindex="-1"></a>ref_dir <span class="sc">|&gt;</span> <span class="fu">filePath</span>(sc_dts[<span class="dv">1</span>]) <span class="sc">|&gt;</span> <span class="fu">list.files</span>()</span>
<span id="cb172-113"><a href="#cb172-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-114"><a href="#cb172-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-115"><a href="#cb172-115" aria-hidden="true" tabindex="-1"></a>This is a common export of single-cell data and can be read directly into Seurat using function <span class="in">`Read10X`</span>.</span>
<span id="cb172-116"><a href="#cb172-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-117"><a href="#cb172-117" aria-hidden="true" tabindex="-1"></a>This snippet will read all of our samples at once:</span>
<span id="cb172-118"><a href="#cb172-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-121"><a href="#cb172-121" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-122"><a href="#cb172-122" aria-hidden="true" tabindex="-1"></a>read_exp_sample <span class="ot">&lt;-</span> <span class="cf">function</span>(sc_file) {</span>
<span id="cb172-123"><a href="#cb172-123" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">'Reading file: '</span>, sc_file)</span>
<span id="cb172-124"><a href="#cb172-124" aria-hidden="true" tabindex="-1"></a>  so <span class="ot">&lt;-</span> ref_dir <span class="sc">|&gt;</span> <span class="fu">filePath</span>(sc_file) <span class="sc">|&gt;</span> <span class="fu">Read10X</span>()</span>
<span id="cb172-125"><a href="#cb172-125" aria-hidden="true" tabindex="-1"></a>  so <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(so, <span class="at">project =</span> <span class="st">"BLCA_NKG2A"</span>)</span>
<span id="cb172-126"><a href="#cb172-126" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"sample"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(sc_file, <span class="st">"^.{4}"</span>)</span>
<span id="cb172-127"><a href="#cb172-127" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"patient"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(sc_file, <span class="st">"Patient.."</span>)</span>
<span id="cb172-128"><a href="#cb172-128" aria-hidden="true" tabindex="-1"></a>  so[[<span class="st">"CD45"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(sc_file, <span class="st">"(?&lt;=CD45).*"</span>)</span>
<span id="cb172-129"><a href="#cb172-129" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(so)</span>
<span id="cb172-130"><a href="#cb172-130" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb172-131"><a href="#cb172-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-132"><a href="#cb172-132" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sc_dts, read_exp_sample)</span>
<span id="cb172-133"><a href="#cb172-133" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sc_data) <span class="ot">&lt;-</span> sc_dts</span>
<span id="cb172-134"><a href="#cb172-134" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-135"><a href="#cb172-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-136"><a href="#cb172-136" aria-hidden="true" tabindex="-1"></a><span class="fu">### Merge samples</span></span>
<span id="cb172-137"><a href="#cb172-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-138"><a href="#cb172-138" aria-hidden="true" tabindex="-1"></a>The script we ran generated 10 Seurat objects, which we will merge into one for further analysis. Each count matrix, at first, will be stored as a <span class="in">`Layer`</span> in our Seurat object.</span>
<span id="cb172-139"><a href="#cb172-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-142"><a href="#cb172-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-143"><a href="#cb172-143" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(sc_data[[<span class="dv">1</span>]], sc_data[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(sc_data)])</span>
<span id="cb172-144"><a href="#cb172-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-145"><a href="#cb172-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-146"><a href="#cb172-146" aria-hidden="true" tabindex="-1"></a><span class="fu">## Seurat syntax basics for object accession</span></span>
<span id="cb172-147"><a href="#cb172-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-148"><a href="#cb172-148" aria-hidden="true" tabindex="-1"></a>Here are some "essential commands" for accession of information inside the Seurat object :<span class="ot">&lt;https://satijalab.org/seurat/articles/essential_commands.html&gt;</span>.</span>
<span id="cb172-149"><a href="#cb172-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-150"><a href="#cb172-150" aria-hidden="true" tabindex="-1"></a>The document also outlines the "basic pipeline".</span>
<span id="cb172-151"><a href="#cb172-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-152"><a href="#cb172-152" aria-hidden="true" tabindex="-1"></a>To see some general info about the object, we can just print it:</span>
<span id="cb172-153"><a href="#cb172-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-156"><a href="#cb172-156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-157"><a href="#cb172-157" aria-hidden="true" tabindex="-1"></a>sc_data</span>
<span id="cb172-158"><a href="#cb172-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-159"><a href="#cb172-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-160"><a href="#cb172-160" aria-hidden="true" tabindex="-1"></a>Note the layers, for the different samples. We can also see what analyses have been run in this object (none yet).</span>
<span id="cb172-161"><a href="#cb172-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-162"><a href="#cb172-162" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Cell metadata</span></span>
<span id="cb172-163"><a href="#cb172-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-164"><a href="#cb172-164" aria-hidden="true" tabindex="-1"></a><span class="in">`[[]]`</span> allows us to access the cell metadata in a concise manner:</span>
<span id="cb172-165"><a href="#cb172-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-168"><a href="#cb172-168" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-169"><a href="#cb172-169" aria-hidden="true" tabindex="-1"></a>sc_data[[]] <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb172-170"><a href="#cb172-170" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-171"><a href="#cb172-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-172"><a href="#cb172-172" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Assay matrix (e.g. counts)</span></span>
<span id="cb172-173"><a href="#cb172-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-174"><a href="#cb172-174" aria-hidden="true" tabindex="-1"></a>To get matrices from the Seurat objects, there are many "official" ways, but one of these two, which have exactly the same output, should work well:</span>
<span id="cb172-175"><a href="#cb172-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-178"><a href="#cb172-178" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-179"><a href="#cb172-179" aria-hidden="true" tabindex="-1"></a><span class="fu">GetAssayData</span>(sc_data, <span class="at">layer =</span> <span class="st">"counts.1"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-180"><a href="#cb172-180" aria-hidden="true" tabindex="-1"></a>sc_data[[<span class="st">"RNA"</span>]]<span class="sc">$</span>counts<span class="fl">.1</span>[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-181"><a href="#cb172-181" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-182"><a href="#cb172-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-183"><a href="#cb172-183" aria-hidden="true" tabindex="-1"></a><span class="fu">### UMI ids</span></span>
<span id="cb172-184"><a href="#cb172-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-185"><a href="#cb172-185" aria-hidden="true" tabindex="-1"></a>To get the "ids" of the cells, we can simply call:</span>
<span id="cb172-186"><a href="#cb172-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-189"><a href="#cb172-189" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-190"><a href="#cb172-190" aria-hidden="true" tabindex="-1"></a><span class="fu">Cells</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-191"><a href="#cb172-191" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-192"><a href="#cb172-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-193"><a href="#cb172-193" aria-hidden="true" tabindex="-1"></a><span class="fu">### Features</span></span>
<span id="cb172-194"><a href="#cb172-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-195"><a href="#cb172-195" aria-hidden="true" tabindex="-1"></a>For a single-cell RNA-seq object, we can see the genes (features) like so:</span>
<span id="cb172-196"><a href="#cb172-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-199"><a href="#cb172-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-200"><a href="#cb172-200" aria-hidden="true" tabindex="-1"></a><span class="fu">Features</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-201"><a href="#cb172-201" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-202"><a href="#cb172-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-203"><a href="#cb172-203" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pre-processing workflow</span></span>
<span id="cb172-204"><a href="#cb172-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-205"><a href="#cb172-205" aria-hidden="true" tabindex="-1"></a>There are very standard steps for pre-processing, but that require some level of decision on the part of the user. The basic tutorial of Seurat, on PBMCs, covers this in detail: https://satijalab.org/seurat/articles/pbmc3k_tutorial</span>
<span id="cb172-206"><a href="#cb172-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-207"><a href="#cb172-207" aria-hidden="true" tabindex="-1"></a><span class="fu">### QC and selecting cells for further analysis</span></span>
<span id="cb172-208"><a href="#cb172-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-209"><a href="#cb172-209" aria-hidden="true" tabindex="-1"></a>This command will figure out the percentage of reads of mitchondrial RNA:</span>
<span id="cb172-210"><a href="#cb172-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-213"><a href="#cb172-213" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-214"><a href="#cb172-214" aria-hidden="true" tabindex="-1"></a>sc_data[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sc_data, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb172-215"><a href="#cb172-215" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-216"><a href="#cb172-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-217"><a href="#cb172-217" aria-hidden="true" tabindex="-1"></a>There are three major pieces of information we'd like to check to make filtering decisions: number of expressed features, number of reads and the percentage of mitochondrial RNA. Here, I plot per sample:</span>
<span id="cb172-218"><a href="#cb172-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-219"><a href="#cb172-219" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=10, fig.height=4}</span></span>
<span id="cb172-220"><a href="#cb172-220" aria-hidden="true" tabindex="-1"></a><span class="in">(VlnPlot(sc_data, </span></span>
<span id="cb172-221"><a href="#cb172-221" aria-hidden="true" tabindex="-1"></a><span class="in">        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), </span></span>
<span id="cb172-222"><a href="#cb172-222" aria-hidden="true" tabindex="-1"></a><span class="in">        group.by = "sample",</span></span>
<span id="cb172-223"><a href="#cb172-223" aria-hidden="true" tabindex="-1"></a><span class="in">        alpha = 0.3, </span></span>
<span id="cb172-224"><a href="#cb172-224" aria-hidden="true" tabindex="-1"></a><span class="in">        ncol = 3) &amp;</span></span>
<span id="cb172-225"><a href="#cb172-225" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean) +</span></span>
<span id="cb172-226"><a href="#cb172-226" aria-hidden="true" tabindex="-1"></a><span class="in">  plot_layout(guides = "collect")</span></span>
<span id="cb172-227"><a href="#cb172-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-228"><a href="#cb172-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-229"><a href="#cb172-229" aria-hidden="true" tabindex="-1"></a>We can also plot this for a particular known variable. For example, here I show for whether the cells were FACS-sorted as CD45 positive or negative:</span>
<span id="cb172-230"><a href="#cb172-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-231"><a href="#cb172-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=10, fig.height=4}</span></span>
<span id="cb172-232"><a href="#cb172-232" aria-hidden="true" tabindex="-1"></a><span class="in">(VlnPlot(sc_data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), </span></span>
<span id="cb172-233"><a href="#cb172-233" aria-hidden="true" tabindex="-1"></a><span class="in">         alpha = 0.3,</span></span>
<span id="cb172-234"><a href="#cb172-234" aria-hidden="true" tabindex="-1"></a><span class="in">         group.by = "CD45", ncol = 3) &amp;</span></span>
<span id="cb172-235"><a href="#cb172-235" aria-hidden="true" tabindex="-1"></a><span class="in">  stat_compare_means() &amp;</span></span>
<span id="cb172-236"><a href="#cb172-236" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean) +</span></span>
<span id="cb172-237"><a href="#cb172-237" aria-hidden="true" tabindex="-1"></a><span class="in">  plot_layout(guides = "collect") +</span></span>
<span id="cb172-238"><a href="#cb172-238" aria-hidden="true" tabindex="-1"></a><span class="in">  plot_annotation(title = "CD45")</span></span>
<span id="cb172-239"><a href="#cb172-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-240"><a href="#cb172-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-241"><a href="#cb172-241" aria-hidden="true" tabindex="-1"></a><span class="fu">### Filtering and Normalization</span></span>
<span id="cb172-242"><a href="#cb172-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-243"><a href="#cb172-243" aria-hidden="true" tabindex="-1"></a>We'll reproduce the filtering and normalization steps proposed by the paper as closely as possible:</span>
<span id="cb172-244"><a href="#cb172-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-245"><a href="#cb172-245" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; For each sample, cells were first selected as expressing less than 16–20% mitochondrial genes and displaying a minimum of 200–300 and a maximum of 2500–3500 features.</span></span>
<span id="cb172-246"><a href="#cb172-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-247"><a href="#cb172-247" aria-hidden="true" tabindex="-1"></a>As we're standardizing pre-processing for all samples, we'll take the least stringent:</span>
<span id="cb172-248"><a href="#cb172-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-251"><a href="#cb172-251" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-252"><a href="#cb172-252" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">subset</span>(sc_data, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span> <span class="sc">&amp;</span> nFeature_RNA <span class="sc">&lt;</span> <span class="dv">3500</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">20</span>)</span>
<span id="cb172-253"><a href="#cb172-253" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-254"><a href="#cb172-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-255"><a href="#cb172-255" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Data were then log-normalized using a scale factor of 10,000.</span></span>
<span id="cb172-256"><a href="#cb172-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-259"><a href="#cb172-259" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-260"><a href="#cb172-260" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sc_data, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="fl">1e4</span>)</span>
<span id="cb172-261"><a href="#cb172-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-262"><a href="#cb172-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-263"><a href="#cb172-263" aria-hidden="true" tabindex="-1"></a><span class="fu">### Feature selection and PCA</span></span>
<span id="cb172-264"><a href="#cb172-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-265"><a href="#cb172-265" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; The 2,000 most variable features were then identified, data were scaled based on all the features </span><span class="sc">\[</span><span class="at">...</span><span class="sc">\]</span></span>
<span id="cb172-266"><a href="#cb172-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-269"><a href="#cb172-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-270"><a href="#cb172-270" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(sc_data, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb172-271"><a href="#cb172-271" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(sc_data, <span class="at">features =</span> <span class="fu">rownames</span>(sc_data))</span>
<span id="cb172-272"><a href="#cb172-272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-273"><a href="#cb172-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-274"><a href="#cb172-274" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="sc">\[</span><span class="at">...</span><span class="sc">\]</span><span class="at"> and principal component analysis was performed.</span></span>
<span id="cb172-275"><a href="#cb172-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-278"><a href="#cb172-278" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-279"><a href="#cb172-279" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(sc_data)</span>
<span id="cb172-280"><a href="#cb172-280" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb172-281"><a href="#cb172-281" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-282"><a href="#cb172-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-283"><a href="#cb172-283" aria-hidden="true" tabindex="-1"></a>We can visualize the PCA with <span class="in">`DimPlot`</span>, coloring by different categorical variables. Here, we see the influence of both the sample (batch) and CD45 positive vs negative (which is correlated to batch but might also contribute to variance).</span>
<span id="cb172-284"><a href="#cb172-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-285"><a href="#cb172-285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=8, fig.height = 4}</span></span>
<span id="cb172-286"><a href="#cb172-286" aria-hidden="true" tabindex="-1"></a><span class="in">DimPlot(sc_data, group.by = c("sample", "CD45"), reduction = "pca") &amp;</span></span>
<span id="cb172-287"><a href="#cb172-287" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean</span></span>
<span id="cb172-288"><a href="#cb172-288" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-289"><a href="#cb172-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-290"><a href="#cb172-290" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Dimensionality determination</span></span>
<span id="cb172-291"><a href="#cb172-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-292"><a href="#cb172-292" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Dimensionality of the dataset was then assessed using the JackStraw and ElbowPlot functions.</span></span>
<span id="cb172-293"><a href="#cb172-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-294"><a href="#cb172-294" aria-hidden="true" tabindex="-1"></a>From Seurat documentation, we see that JackStraw is quite computationally intensive:</span>
<span id="cb172-295"><a href="#cb172-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-296"><a href="#cb172-296" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; In Macosko et al, we implemented a resampling test inspired by the JackStraw procedure. While still available in Seurat (see previous vignette), this is a slow and computationally expensive procedure, and we is no longer routinely used in single cell analysis.</span></span>
<span id="cb172-297"><a href="#cb172-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-298"><a href="#cb172-298" aria-hidden="true" tabindex="-1"></a>This is not the most beautiful Elbow plot ever, but the idea is to find the "inflection point" of the plot (I'm estimating at about 10) and go a little bit further into the "stable" area, where adding more dimensions doesn't add more infromation.</span>
<span id="cb172-299"><a href="#cb172-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-300"><a href="#cb172-300" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width = 5, fig.height=3}</span></span>
<span id="cb172-301"><a href="#cb172-301" aria-hidden="true" tabindex="-1"></a><span class="in">ElbowPlot(sc_data, ndims = 30)</span></span>
<span id="cb172-302"><a href="#cb172-302" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-303"><a href="#cb172-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-304"><a href="#cb172-304" aria-hidden="true" tabindex="-1"></a>For methods that require the dimensionality, we'll set it at 15.</span>
<span id="cb172-305"><a href="#cb172-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-306"><a href="#cb172-306" aria-hidden="true" tabindex="-1"></a><span class="fu">### Doublet detection</span></span>
<span id="cb172-307"><a href="#cb172-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-308"><a href="#cb172-308" aria-hidden="true" tabindex="-1"></a>It's never a bad idea to add some doublet detection.</span>
<span id="cb172-309"><a href="#cb172-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-310"><a href="#cb172-310" aria-hidden="true" tabindex="-1"></a>Doublets are UMIs that contained more than one cell and thefore have a mixed profile.</span>
<span id="cb172-311"><a href="#cb172-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-312"><a href="#cb172-312" aria-hidden="true" tabindex="-1"></a>This benchmark shows some of the main methods (though new ones may emerge): https://www.sciencedirect.com/science/article/pii/S2405471220304592</span>
<span id="cb172-313"><a href="#cb172-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-314"><a href="#cb172-314" aria-hidden="true" tabindex="-1"></a>We'll use cxds from scdc because it's fast and reasonably accurate.</span>
<span id="cb172-315"><a href="#cb172-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-316"><a href="#cb172-316" aria-hidden="true" tabindex="-1"></a>However, it uses a <span class="in">`SingleCellExperiment`</span> standard. Luckily, we can easily convert between <span class="in">`SeuratObject`</span> and <span class="in">`SingleCellExperiment`</span>.</span>
<span id="cb172-317"><a href="#cb172-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-320"><a href="#cb172-320" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-321"><a href="#cb172-321" aria-hidden="true" tabindex="-1"></a>cxds_scores <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">SplitObject</span>(sc_data, <span class="at">split.by =</span> <span class="st">"sample"</span>), <span class="cf">function</span>(samp_sc) {</span>
<span id="cb172-322"><a href="#cb172-322" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> samp_sc <span class="sc">|&gt;</span></span>
<span id="cb172-323"><a href="#cb172-323" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.SingleCellExperiment</span>() <span class="sc">|&gt;</span></span>
<span id="cb172-324"><a href="#cb172-324" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cxds</span>()</span>
<span id="cb172-325"><a href="#cb172-325" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce<span class="sc">$</span>cxds_score)</span>
<span id="cb172-326"><a href="#cb172-326" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb172-327"><a href="#cb172-327" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb172-328"><a href="#cb172-328" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cxds_scores) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb172-329"><a href="#cb172-329" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>cxds_scores <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, cxds_scores)[<span class="fu">Cells</span>(sc_data)]</span>
<span id="cb172-330"><a href="#cb172-330" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-331"><a href="#cb172-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-332"><a href="#cb172-332" aria-hidden="true" tabindex="-1"></a><span class="fu">## Clustering</span></span>
<span id="cb172-333"><a href="#cb172-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-334"><a href="#cb172-334" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Clusters were calculated </span><span class="sc">\[</span><span class="at">...</span><span class="sc">\]</span></span>
<span id="cb172-335"><a href="#cb172-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-336"><a href="#cb172-336" aria-hidden="true" tabindex="-1"></a>Before clustering, we need to establish a neighborhood graph and compute shared nearest neighbors (SNN). We'll do this on 15 PCs:</span>
<span id="cb172-337"><a href="#cb172-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-340"><a href="#cb172-340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-341"><a href="#cb172-341" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sc_data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb172-342"><a href="#cb172-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-343"><a href="#cb172-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-344"><a href="#cb172-344" aria-hidden="true" tabindex="-1"></a>From this, we can use "FindClusters" for Louvain clustering, which uses a rather temperamental "resolution" parameter, very dependent on datasets usually.</span>
<span id="cb172-345"><a href="#cb172-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-346"><a href="#cb172-346" aria-hidden="true" tabindex="-1"></a>The largest the "resolution", the more clusters are obtained.</span>
<span id="cb172-347"><a href="#cb172-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-350"><a href="#cb172-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-351"><a href="#cb172-351" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sc_data, <span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb172-352"><a href="#cb172-352" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-353"><a href="#cb172-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-354"><a href="#cb172-354" aria-hidden="true" tabindex="-1"></a>UMAP and t-SNE are dimensionality reductions for visualization. They are very popular and we'll use them widely. They minimize local distances, often making "neat" looking clusters.</span>
<span id="cb172-355"><a href="#cb172-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-356"><a href="#cb172-356" aria-hidden="true" tabindex="-1"></a>Be wary of not using the UMAP projection itself as an input for analysis, as they contain lots of data distortions (which is normal of anything that is reducing a space of 3000 features to 2 dimensions)</span>
<span id="cb172-357"><a href="#cb172-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-358"><a href="#cb172-358" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="sc">\[</span><span class="at">...</span><span class="sc">\]</span><span class="at"> and data dimensions were reduced using the t-SNE and UMAP methods.</span></span>
<span id="cb172-359"><a href="#cb172-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-362"><a href="#cb172-362" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-363"><a href="#cb172-363" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb172-364"><a href="#cb172-364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-365"><a href="#cb172-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-366"><a href="#cb172-366" aria-hidden="true" tabindex="-1"></a>We can check the state of our Seurat object:</span>
<span id="cb172-367"><a href="#cb172-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-370"><a href="#cb172-370" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-371"><a href="#cb172-371" aria-hidden="true" tabindex="-1"></a>sc_data</span>
<span id="cb172-372"><a href="#cb172-372" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-373"><a href="#cb172-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-374"><a href="#cb172-374" aria-hidden="true" tabindex="-1"></a>And plot the Seurat clusters... only to find we're definitely clustering per sample.</span>
<span id="cb172-375"><a href="#cb172-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-376"><a href="#cb172-376" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width = 12, fig.height=5}</span></span>
<span id="cb172-377"><a href="#cb172-377" aria-hidden="true" tabindex="-1"></a><span class="in">DimPlot(sc_data, group.by = c("unintegrated_clusters", "sample"), reduction = "umap", alpha = 0.3) &amp; </span></span>
<span id="cb172-378"><a href="#cb172-378" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean</span></span>
<span id="cb172-379"><a href="#cb172-379" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-380"><a href="#cb172-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-381"><a href="#cb172-381" aria-hidden="true" tabindex="-1"></a><span class="fu">## Integration</span></span>
<span id="cb172-382"><a href="#cb172-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-383"><a href="#cb172-383" aria-hidden="true" tabindex="-1"></a>Batch effect is rampant in single-cell RNA-seq and there are lots of integration methods proposed to deal with this. They often return a dimensionality reduction (more precisely, latent space) that minimizes this batch effect.</span>
<span id="cb172-384"><a href="#cb172-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-385"><a href="#cb172-385" aria-hidden="true" tabindex="-1"></a>There are 5 methods implemented in Seurat alone. We'll use Harmony, because it's fairly fast and effective.</span>
<span id="cb172-386"><a href="#cb172-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-387"><a href="#cb172-387" aria-hidden="true" tabindex="-1"></a>See the reference tutorial: https://satijalab.org/seurat/articles/seurat5_integration</span>
<span id="cb172-388"><a href="#cb172-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-389"><a href="#cb172-389" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Bulk cells were merged for the HLA analysis, while CD8 T cells only were merged for the CD8 T cell analysis. CD8 T cells were identified using the FindAllMarkers and the VlnPlot functions.</span></span>
<span id="cb172-390"><a href="#cb172-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-393"><a href="#cb172-393" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-394"><a href="#cb172-394" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(sc_data, </span>
<span id="cb172-395"><a href="#cb172-395" aria-hidden="true" tabindex="-1"></a>                           <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb172-396"><a href="#cb172-396" aria-hidden="true" tabindex="-1"></a>                           <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb172-397"><a href="#cb172-397" aria-hidden="true" tabindex="-1"></a>                           <span class="at">new.reduction =</span> <span class="st">"harmony"</span>,</span>
<span id="cb172-398"><a href="#cb172-398" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb172-399"><a href="#cb172-399" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb172-400"><a href="#cb172-400" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb172-401"><a href="#cb172-401" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-402"><a href="#cb172-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-403"><a href="#cb172-403" aria-hidden="true" tabindex="-1"></a>We'll need re-run clustering algorithm in the Harmony latent space (here, using 30 dimensions)</span>
<span id="cb172-404"><a href="#cb172-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-407"><a href="#cb172-407" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-408"><a href="#cb172-408" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sc_data, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb172-409"><a href="#cb172-409" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sc_data, <span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">cluster.name =</span> <span class="st">"harmony_clusters"</span>)</span>
<span id="cb172-410"><a href="#cb172-410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-411"><a href="#cb172-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-412"><a href="#cb172-412" aria-hidden="true" tabindex="-1"></a>We can also run a UMAP in this latent space:</span>
<span id="cb172-413"><a href="#cb172-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-416"><a href="#cb172-416" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-417"><a href="#cb172-417" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction.name =</span> <span class="st">"umap_harmony"</span>)</span>
<span id="cb172-418"><a href="#cb172-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-419"><a href="#cb172-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-420"><a href="#cb172-420" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width = 12, fig.height=7}</span></span>
<span id="cb172-421"><a href="#cb172-421" aria-hidden="true" tabindex="-1"></a><span class="in">DimPlot(sc_data, </span></span>
<span id="cb172-422"><a href="#cb172-422" aria-hidden="true" tabindex="-1"></a><span class="in">        group.by = c("unintegrated_clusters", "harmony_clusters", "sample"), </span></span>
<span id="cb172-423"><a href="#cb172-423" aria-hidden="true" tabindex="-1"></a><span class="in">        reduction = "umap_harmony",</span></span>
<span id="cb172-424"><a href="#cb172-424" aria-hidden="true" tabindex="-1"></a><span class="in">        alpha = 0.2,</span></span>
<span id="cb172-425"><a href="#cb172-425" aria-hidden="true" tabindex="-1"></a><span class="in">        ncol = 2) &amp; </span></span>
<span id="cb172-426"><a href="#cb172-426" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean</span></span>
<span id="cb172-427"><a href="#cb172-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-428"><a href="#cb172-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-429"><a href="#cb172-429" aria-hidden="true" tabindex="-1"></a><span class="fu">### Check biases</span></span>
<span id="cb172-430"><a href="#cb172-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-431"><a href="#cb172-431" aria-hidden="true" tabindex="-1"></a>We can check back on our features that could be markers of bias in a cluster, and check back in with our doublet scores:</span>
<span id="cb172-432"><a href="#cb172-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-435"><a href="#cb172-435" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-436"><a href="#cb172-436" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>cxds_scores[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-437"><a href="#cb172-437" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-438"><a href="#cb172-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-439"><a href="#cb172-439" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width = 12}</span></span>
<span id="cb172-440"><a href="#cb172-440" aria-hidden="true" tabindex="-1"></a><span class="in">RidgePlot(sc_data, </span></span>
<span id="cb172-441"><a href="#cb172-441" aria-hidden="true" tabindex="-1"></a><span class="in">        features = c("nCount_RNA", "nFeature_RNA", </span></span>
<span id="cb172-442"><a href="#cb172-442" aria-hidden="true" tabindex="-1"></a><span class="in">                     "percent.mt","cxds_scores"), </span></span>
<span id="cb172-443"><a href="#cb172-443" aria-hidden="true" tabindex="-1"></a><span class="in">        ncol = 2) &amp;</span></span>
<span id="cb172-444"><a href="#cb172-444" aria-hidden="true" tabindex="-1"></a><span class="in">  guides(fill = "none") &amp;</span></span>
<span id="cb172-445"><a href="#cb172-445" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean </span></span>
<span id="cb172-446"><a href="#cb172-446" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-447"><a href="#cb172-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-448"><a href="#cb172-448" aria-hidden="true" tabindex="-1"></a>We can see, for example, that cluster 17 seems to be enriched in doublets.</span>
<span id="cb172-449"><a href="#cb172-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-450"><a href="#cb172-450" aria-hidden="true" tabindex="-1"></a>To compare the co-occurence of two categorical variables, we can do a heatmap:</span>
<span id="cb172-451"><a href="#cb172-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-454"><a href="#cb172-454" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-455"><a href="#cb172-455" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sc_data<span class="sc">$</span>harmony_clusters, sc_data<span class="sc">$</span>sample) <span class="sc">|&gt;</span></span>
<span id="cb172-456"><a href="#cb172-456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pheatmap</span>(<span class="at">scale =</span> <span class="st">"row"</span>)</span>
<span id="cb172-457"><a href="#cb172-457" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-458"><a href="#cb172-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-459"><a href="#cb172-459" aria-hidden="true" tabindex="-1"></a>From this, we can exclude cluster 17:</span>
<span id="cb172-460"><a href="#cb172-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-463"><a href="#cb172-463" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-464"><a href="#cb172-464" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">subset</span>(sc_data, <span class="at">subset =</span> harmony_clusters <span class="sc">!=</span> <span class="dv">17</span>)</span>
<span id="cb172-465"><a href="#cb172-465" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-466"><a href="#cb172-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-467"><a href="#cb172-467" aria-hidden="true" tabindex="-1"></a><span class="fu">## Annotation</span></span>
<span id="cb172-468"><a href="#cb172-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-469"><a href="#cb172-469" aria-hidden="true" tabindex="-1"></a><span class="fu">### Differential Expression</span></span>
<span id="cb172-470"><a href="#cb172-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-471"><a href="#cb172-471" aria-hidden="true" tabindex="-1"></a>Differential expression analysis will by default run on the "identity" feature, which we make sure to set as "harmony_clusters" at this point:</span>
<span id="cb172-472"><a href="#cb172-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-475"><a href="#cb172-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-476"><a href="#cb172-476" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(sc_data) <span class="ot">&lt;-</span> sc_data<span class="sc">$</span>harmony_clusters</span>
<span id="cb172-477"><a href="#cb172-477" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(sc_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb172-478"><a href="#cb172-478" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-479"><a href="#cb172-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-480"><a href="#cb172-480" aria-hidden="true" tabindex="-1"></a>For differential analysis to use all layers at the same time, we need to join them before finding markers (each clusters vs all others):</span>
<span id="cb172-481"><a href="#cb172-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-484"><a href="#cb172-484" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-485"><a href="#cb172-485" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(sc_data)</span>
<span id="cb172-486"><a href="#cb172-486" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(sc_data)</span>
<span id="cb172-487"><a href="#cb172-487" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-488"><a href="#cb172-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-489"><a href="#cb172-489" aria-hidden="true" tabindex="-1"></a>For annotation, we can now use this table:</span>
<span id="cb172-490"><a href="#cb172-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-493"><a href="#cb172-493" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-494"><a href="#cb172-494" aria-hidden="true" tabindex="-1"></a>markers <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb172-495"><a href="#cb172-495" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-496"><a href="#cb172-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-497"><a href="#cb172-497" aria-hidden="true" tabindex="-1"></a>And visualize either top markers or any chosen features in many different ways:</span>
<span id="cb172-498"><a href="#cb172-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-501"><a href="#cb172-501" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-502"><a href="#cb172-502" aria-hidden="true" tabindex="-1"></a>top_markers <span class="ot">&lt;-</span> markers <span class="sc">|&gt;</span></span>
<span id="cb172-503"><a href="#cb172-503" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">|&gt;</span></span>
<span id="cb172-504"><a href="#cb172-504" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb172-505"><a href="#cb172-505" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(p_val_adj) <span class="sc">|&gt;</span></span>
<span id="cb172-506"><a href="#cb172-506" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb172-507"><a href="#cb172-507" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb172-508"><a href="#cb172-508" aria-hidden="true" tabindex="-1"></a>top5 <span class="ot">&lt;-</span> <span class="fu">unique</span>(top_markers<span class="sc">$</span>gene)</span>
<span id="cb172-509"><a href="#cb172-509" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-510"><a href="#cb172-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-511"><a href="#cb172-511" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Heatmap</span></span>
<span id="cb172-512"><a href="#cb172-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-513"><a href="#cb172-513" aria-hidden="true" tabindex="-1"></a>The heatmap takes a while to run, but provides a comprehensive view of a marker expression over the cells:</span>
<span id="cb172-514"><a href="#cb172-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-515"><a href="#cb172-515" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width = 10, fig.heigth = 10}</span></span>
<span id="cb172-516"><a href="#cb172-516" aria-hidden="true" tabindex="-1"></a><span class="in">DoHeatmap(sc_data, features = top5, raster = TRUE) +</span></span>
<span id="cb172-517"><a href="#cb172-517" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_fill_viridis_c()</span></span>
<span id="cb172-518"><a href="#cb172-518" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-519"><a href="#cb172-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-520"><a href="#cb172-520" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Dot plot</span></span>
<span id="cb172-521"><a href="#cb172-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-522"><a href="#cb172-522" aria-hidden="true" tabindex="-1"></a>The dotplot summarizes the information into the heatmap and gives the same importance to all clusters, independent of size:</span>
<span id="cb172-523"><a href="#cb172-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-524"><a href="#cb172-524" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=15}</span></span>
<span id="cb172-525"><a href="#cb172-525" aria-hidden="true" tabindex="-1"></a><span class="in">DotPlot(sc_data, features = top5, cluster.idents = TRUE) + theme_clean</span></span>
<span id="cb172-526"><a href="#cb172-526" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-527"><a href="#cb172-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-528"><a href="#cb172-528" aria-hidden="true" tabindex="-1"></a><span class="fu">#### FeaturePlot</span></span>
<span id="cb172-529"><a href="#cb172-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-530"><a href="#cb172-530" aria-hidden="true" tabindex="-1"></a><span class="in">`FeaturePlot`</span> is the same as <span class="in">`DimPlot`</span>, but using one or many continuous values instead:</span>
<span id="cb172-531"><a href="#cb172-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-532"><a href="#cb172-532" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=8, fig.height=12}</span></span>
<span id="cb172-533"><a href="#cb172-533" aria-hidden="true" tabindex="-1"></a><span class="in">FeaturePlot(sc_data, reduction = "umap_harmony",</span></span>
<span id="cb172-534"><a href="#cb172-534" aria-hidden="true" tabindex="-1"></a><span class="in">            features = c("CD3D", "MS4A1",</span></span>
<span id="cb172-535"><a href="#cb172-535" aria-hidden="true" tabindex="-1"></a><span class="in">                         "CD14", "PECAM1",</span></span>
<span id="cb172-536"><a href="#cb172-536" aria-hidden="true" tabindex="-1"></a><span class="in">                         "KRT20", "KRT5")) &amp;</span></span>
<span id="cb172-537"><a href="#cb172-537" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_void()</span></span>
<span id="cb172-538"><a href="#cb172-538" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-539"><a href="#cb172-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-540"><a href="#cb172-540" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reference-based</span></span>
<span id="cb172-541"><a href="#cb172-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-542"><a href="#cb172-542" aria-hidden="true" tabindex="-1"></a>There are an infinity of methods for cell type annotation. I've found that using a good reference is more important than the method for annotation.</span>
<span id="cb172-543"><a href="#cb172-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-544"><a href="#cb172-544" aria-hidden="true" tabindex="-1"></a>We'll show reference-based annotation using SingleR, mostly because it's easy to use but also well performing.</span>
<span id="cb172-545"><a href="#cb172-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-546"><a href="#cb172-546" aria-hidden="true" tabindex="-1"></a><span class="fu">#### "Built-in" references</span></span>
<span id="cb172-547"><a href="#cb172-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-548"><a href="#cb172-548" aria-hidden="true" tabindex="-1"></a>Every automatic annotation method uses a reference. <span class="in">`celldex`</span> provides an easy way to access some annotation sources, notably the Human Primary Cell Atlas, which we will use in this example.</span>
<span id="cb172-549"><a href="#cb172-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-552"><a href="#cb172-552" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-553"><a href="#cb172-553" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(celldex)</span>
<span id="cb172-554"><a href="#cb172-554" aria-hidden="true" tabindex="-1"></a><span class="fu">surveyReferences</span>() <span class="sc">|&gt;</span>  <span class="fu">as_tibble</span>()</span>
<span id="cb172-555"><a href="#cb172-555" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-556"><a href="#cb172-556" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-559"><a href="#cb172-559" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-560"><a href="#cb172-560" aria-hidden="true" tabindex="-1"></a>hpca_se <span class="ot">&lt;-</span> celldex<span class="sc">::</span><span class="fu">HumanPrimaryCellAtlasData</span>()</span>
<span id="cb172-561"><a href="#cb172-561" aria-hidden="true" tabindex="-1"></a>hpca_se</span>
<span id="cb172-562"><a href="#cb172-562" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-563"><a href="#cb172-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-564"><a href="#cb172-564" aria-hidden="true" tabindex="-1"></a>It's given as a <span class="in">`SummarizedExperiment`</span> of purified bulk tumors.=, annotated as follows:</span>
<span id="cb172-565"><a href="#cb172-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-568"><a href="#cb172-568" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-569"><a href="#cb172-569" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(hpca_se<span class="sc">$</span>label.main)</span>
<span id="cb172-570"><a href="#cb172-570" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-571"><a href="#cb172-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-572"><a href="#cb172-572" aria-hidden="true" tabindex="-1"></a>There's also a finer annotation, which we won't use:</span>
<span id="cb172-573"><a href="#cb172-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-576"><a href="#cb172-576" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-577"><a href="#cb172-577" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(hpca_se<span class="sc">$</span>label.fine) <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">20</span>)</span>
<span id="cb172-578"><a href="#cb172-578" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-579"><a href="#cb172-579" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-580"><a href="#cb172-580" aria-hidden="true" tabindex="-1"></a>SingleR is really simple to use, we just need to provide an expression matrix and reference as <span class="in">`SummarizedExperiment`</span>:</span>
<span id="cb172-581"><a href="#cb172-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-584"><a href="#cb172-584" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-585"><a href="#cb172-585" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleR)</span>
<span id="cb172-586"><a href="#cb172-586" aria-hidden="true" tabindex="-1"></a>pred_hpca <span class="ot">&lt;-</span> <span class="fu">SingleR</span>(<span class="at">test =</span> <span class="fu">GetAssayData</span>(sc_data), <span class="at">ref =</span> hpca_se, <span class="at">assay.type.test=</span><span class="dv">1</span>,</span>
<span id="cb172-587"><a href="#cb172-587" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> hpca_se<span class="sc">$</span>label.main)</span>
<span id="cb172-588"><a href="#cb172-588" aria-hidden="true" tabindex="-1"></a>pred_hpca</span>
<span id="cb172-589"><a href="#cb172-589" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-590"><a href="#cb172-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-591"><a href="#cb172-591" aria-hidden="true" tabindex="-1"></a>Once we have the predictions, I add it to the Seurat object:</span>
<span id="cb172-592"><a href="#cb172-592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-595"><a href="#cb172-595" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-596"><a href="#cb172-596" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">@</span>tools[[<span class="st">"SingleR"</span>]] <span class="ot">&lt;-</span> pred_hpca <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb172-597"><a href="#cb172-597" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>singleR_pred <span class="ot">&lt;-</span> pred_hpca<span class="sc">$</span>pruned.labels</span>
<span id="cb172-598"><a href="#cb172-598" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-599"><a href="#cb172-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-600"><a href="#cb172-600" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Harmonizing automatic predictions</span></span>
<span id="cb172-601"><a href="#cb172-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-602"><a href="#cb172-602" aria-hidden="true" tabindex="-1"></a>We can check the prediction results and "harmonize" them by cluster, i.e. take the cluster-level mode of the annotation:</span>
<span id="cb172-603"><a href="#cb172-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-606"><a href="#cb172-606" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-607"><a href="#cb172-607" aria-hidden="true" tabindex="-1"></a>harmony_singler <span class="ot">&lt;-</span> <span class="fu">table</span>(sc_data<span class="sc">$</span>singleR_pred, <span class="fu">factor</span>(sc_data<span class="sc">$</span>harmony_clusters))</span>
<span id="cb172-608"><a href="#cb172-608" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(harmony_singler, <span class="at">scale =</span> <span class="st">"column"</span>)</span>
<span id="cb172-609"><a href="#cb172-609" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-610"><a href="#cb172-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-613"><a href="#cb172-613" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-614"><a href="#cb172-614" aria-hidden="true" tabindex="-1"></a>harmony_map1 <span class="ot">&lt;-</span> <span class="fu">apply</span>(harmony_singler, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">names</span>(<span class="fu">which.max</span>(x)))</span>
<span id="cb172-615"><a href="#cb172-615" aria-hidden="true" tabindex="-1"></a>harmony_map1 <span class="ot">&lt;-</span> <span class="fu">structure</span>(<span class="fu">names</span>(harmony_map1), <span class="at">names =</span> harmony_map1)</span>
<span id="cb172-616"><a href="#cb172-616" aria-hidden="true" tabindex="-1"></a>harmony_map1</span>
<span id="cb172-617"><a href="#cb172-617" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-618"><a href="#cb172-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-621"><a href="#cb172-621" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-622"><a href="#cb172-622" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">$</span>singleR_clean <span class="ot">&lt;-</span> <span class="fu">fct_recode</span>(sc_data<span class="sc">$</span>harmony_clusters, <span class="sc">!!!</span> harmony_map1)</span>
<span id="cb172-623"><a href="#cb172-623" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-624"><a href="#cb172-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-625"><a href="#cb172-625" aria-hidden="true" tabindex="-1"></a>If we plot, we can see the difference between the "pred" and "clean":</span>
<span id="cb172-626"><a href="#cb172-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-627"><a href="#cb172-627" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=12, fig.height=7}</span></span>
<span id="cb172-628"><a href="#cb172-628" aria-hidden="true" tabindex="-1"></a><span class="in">DimPlot(sc_data, </span></span>
<span id="cb172-629"><a href="#cb172-629" aria-hidden="true" tabindex="-1"></a><span class="in">        group.by = c("harmony_clusters", "singleR_pred", </span></span>
<span id="cb172-630"><a href="#cb172-630" aria-hidden="true" tabindex="-1"></a><span class="in">                     "singleR_clean", "sample"), </span></span>
<span id="cb172-631"><a href="#cb172-631" aria-hidden="true" tabindex="-1"></a><span class="in">        reduction = "umap_harmony",</span></span>
<span id="cb172-632"><a href="#cb172-632" aria-hidden="true" tabindex="-1"></a><span class="in">        ncol = 2,</span></span>
<span id="cb172-633"><a href="#cb172-633" aria-hidden="true" tabindex="-1"></a><span class="in">        alpha = 0.2) &amp; </span></span>
<span id="cb172-634"><a href="#cb172-634" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean &amp;</span></span>
<span id="cb172-635"><a href="#cb172-635" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_color_igv()</span></span>
<span id="cb172-636"><a href="#cb172-636" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-637"><a href="#cb172-637" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-638"><a href="#cb172-638" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Single-cell</span></span>
<span id="cb172-639"><a href="#cb172-639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-640"><a href="#cb172-640" aria-hidden="true" tabindex="-1"></a>Most methods will work best with a relevant single-cell reference.</span>
<span id="cb172-641"><a href="#cb172-641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-642"><a href="#cb172-642" aria-hidden="true" tabindex="-1"></a>Here is a not-so-recent but independent benchmark: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1795-z</span>
<span id="cb172-643"><a href="#cb172-643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-644"><a href="#cb172-644" aria-hidden="true" tabindex="-1"></a>It can be interesting to try to use this tutorial to annotate a query dataset based on a reference atlas: https://satijalab.org/seurat/articles/integration_mapping</span>
<span id="cb172-645"><a href="#cb172-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-646"><a href="#cb172-646" aria-hidden="true" tabindex="-1"></a><span class="fu">## Extras</span></span>
<span id="cb172-647"><a href="#cb172-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-648"><a href="#cb172-648" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sub-cluster</span></span>
<span id="cb172-649"><a href="#cb172-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-650"><a href="#cb172-650" aria-hidden="true" tabindex="-1"></a>In their manuscript, Salomé et al selected the T cells and subclustered them. This is a common strategy when there is a main population of interest. In fact, most of their analysis concentrated on the CD8 T cells.</span>
<span id="cb172-651"><a href="#cb172-651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-652"><a href="#cb172-652" aria-hidden="true" tabindex="-1"></a>At this point, we could take our annotated data, subset the T cells and re-run the pipeline to find new relevant subgroups.</span>
<span id="cb172-653"><a href="#cb172-653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-654"><a href="#cb172-654" aria-hidden="true" tabindex="-1"></a><span class="fu">### Trajectory analysis</span></span>
<span id="cb172-655"><a href="#cb172-655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-656"><a href="#cb172-656" aria-hidden="true" tabindex="-1"></a>Trajectory analysis is used to reconstruct and visualize the dynamic processes of cellular development, differentiation, and state transitions. It infers the order of cells in a particular trajectory, allows us to calculate continuous "pseudotimes".</span>
<span id="cb172-657"><a href="#cb172-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-658"><a href="#cb172-658" aria-hidden="true" tabindex="-1"></a>It can be used to map out potential lineage relationships, identify intermediate cell states, and elucidate the pathways underlying tissue development, immune responses, or disease progression, for example.</span>
<span id="cb172-659"><a href="#cb172-659" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-660"><a href="#cb172-660" aria-hidden="true" tabindex="-1"></a>In Figure 4, Salomé et al used this to show a subset of differentiated PD-1+ TRM CD8+ T Cells that retain independent function.</span>
<span id="cb172-661"><a href="#cb172-661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-662"><a href="#cb172-662" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/Screenshot from 2025-02-13 18-42-25.png)</span></span>
<span id="cb172-663"><a href="#cb172-663" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-664"><a href="#cb172-664" aria-hidden="true" tabindex="-1"></a>A not so new but very comprehensive benchmark of trajectory inference by Saelens et al is a great jumping point: <span class="ot">&lt;https://www.nature.com/articles/s41587-019-0071-9&gt;</span></span>
<span id="cb172-665"><a href="#cb172-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-666"><a href="#cb172-666" aria-hidden="true" tabindex="-1"></a>Accompanied by the superb dynverse wrappers, which provide all methods in the benchmark, as well as guidance on what to choose: <span class="ot">&lt;https://dynverse.org/&gt;</span></span>
<span id="cb172-667"><a href="#cb172-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-668"><a href="#cb172-668" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cell cycle</span></span>
<span id="cb172-669"><a href="#cb172-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-670"><a href="#cb172-670" aria-hidden="true" tabindex="-1"></a>There are many methods to infer cell cycle stage and score cell cycle in Seurat, but it also has a built in one that works quite decently:</span>
<span id="cb172-671"><a href="#cb172-671" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-674"><a href="#cb172-674" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb172-675"><a href="#cb172-675" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(sc_data, </span>
<span id="cb172-676"><a href="#cb172-676" aria-hidden="true" tabindex="-1"></a>                 <span class="at">s.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>s.genes, </span>
<span id="cb172-677"><a href="#cb172-677" aria-hidden="true" tabindex="-1"></a>                 <span class="at">g2m.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>g2m.genes)</span>
<span id="cb172-678"><a href="#cb172-678" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-679"><a href="#cb172-679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-680"><a href="#cb172-680" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=8, fig.height=4}</span></span>
<span id="cb172-681"><a href="#cb172-681" aria-hidden="true" tabindex="-1"></a><span class="in">FeaturePlot(sc_data, </span></span>
<span id="cb172-682"><a href="#cb172-682" aria-hidden="true" tabindex="-1"></a><span class="in">        features = c("S.Score", "G2M.Score"),</span></span>
<span id="cb172-683"><a href="#cb172-683" aria-hidden="true" tabindex="-1"></a><span class="in">        reduction = "umap_harmony",</span></span>
<span id="cb172-684"><a href="#cb172-684" aria-hidden="true" tabindex="-1"></a><span class="in">        ncol = 2,</span></span>
<span id="cb172-685"><a href="#cb172-685" aria-hidden="true" tabindex="-1"></a><span class="in">        alpha = 0.2) &amp; </span></span>
<span id="cb172-686"><a href="#cb172-686" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_clean</span></span>
<span id="cb172-687"><a href="#cb172-687" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb172-688"><a href="#cb172-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-689"><a href="#cb172-689" aria-hidden="true" tabindex="-1"></a>We can see very little cycling except in one "epithelial" (tumor) area we previously annotated.</span>
<span id="cb172-690"><a href="#cb172-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-691"><a href="#cb172-691" aria-hidden="true" tabindex="-1"></a><span class="fu">### RNA velocity</span></span>
<span id="cb172-692"><a href="#cb172-692" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-693"><a href="#cb172-693" aria-hidden="true" tabindex="-1"></a>RNA velocity is a computation method that attempts to predict the future transcriptional state of individual cells. It distinguishes between unspliced (pre-mRNA) and spliced (mature mRNA) transcripts within each cell, and therefore necessitates a special input in which these reads have been separated. This differentiation allows the estimation of the rate at which genes are being transcribed and processed, providing insights into the "direction" and "speed" of cellular state changes.</span>
<span id="cb172-694"><a href="#cb172-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-695"><a href="#cb172-695" aria-hidden="true" tabindex="-1"></a>The primary goal of RNA velocity analysis is to elucidate the trajectories of cellular differentiation, activation, and other dynamic processes, thereby offering a deeper understanding of cellular development and lineage relationships. By forecasting how cells are likely to evolve, it can help in identifying transient cell states, and uncovering mechanisms underlying tissue development, regeneration, and disease progression.</span>
<span id="cb172-696"><a href="#cb172-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-697"><a href="#cb172-697" aria-hidden="true" tabindex="-1"></a>It can be used together with traditional trajectory analysis.</span>
<span id="cb172-698"><a href="#cb172-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-699"><a href="#cb172-699" aria-hidden="true" tabindex="-1"></a>Most RNA velocity tools work in Python, the two most well-known being <span class="in">`velocyto`</span> and <span class="in">`scVelo`</span>. Both have tutorials showing how to use reticulate to run their workflows in R.</span>
<span id="cb172-700"><a href="#cb172-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-701"><a href="#cb172-701" aria-hidden="true" tabindex="-1"></a><span class="fu">### Pathway enrichment</span></span>
<span id="cb172-702"><a href="#cb172-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-703"><a href="#cb172-703" aria-hidden="true" tabindex="-1"></a>Pathway enrichment aims to identify biological pathways, processes, or functional categories that are significantly overrepresented within a set of genes of interest, such as those differentially expressed between distinct cell populations or states.</span>
<span id="cb172-704"><a href="#cb172-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-705"><a href="#cb172-705" aria-hidden="true" tabindex="-1"></a>This analysis typically involves mapping the selected genes to curated pathway databases like Gene Ontology, KEGG, or Reactome and applying statistical methods to assess the significance of their enrichment.</span>
<span id="cb172-706"><a href="#cb172-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-707"><a href="#cb172-707" aria-hidden="true" tabindex="-1"></a>There are many methods to do this, but a easy to use is decoupleR, in which AUCell is implemented, which is a well-performing scoring method: <span class="ot">&lt;https://www.bioconductor.org/packages/release/bioc/vignettes/decoupleR/inst/doc/decoupleR.html&gt;</span></span>
<span id="cb172-708"><a href="#cb172-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-709"><a href="#cb172-709" aria-hidden="true" tabindex="-1"></a><span class="fu">### Gene regulatory network</span></span>
<span id="cb172-710"><a href="#cb172-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-711"><a href="#cb172-711" aria-hidden="true" tabindex="-1"></a>Gene regulatory network (GRN) analysis seeks to identify key transcription factors, regulatory motifs, and signaling pathways that drive cell states, differentiation, and responses to environmental stimuli. The ultimate goal of GRN analysis is to map out the regulatory architecture that underpins cellular heterogeneity, enabling <span class="sc">\\</span> to pinpoint master regulators and understand the molecular mechanisms underlying various biological processes and disease states.</span>
<span id="cb172-712"><a href="#cb172-712" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-713"><a href="#cb172-713" aria-hidden="true" tabindex="-1"></a>From an algorithmic point of view, this involves inferring relationships such as gene co-expression, regulatory influences, and causal interactions.</span>
<span id="cb172-714"><a href="#cb172-714" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-715"><a href="#cb172-715" aria-hidden="true" tabindex="-1"></a>SCENIC is the most popular method to do this, with the Python implementation being much more performant (<span class="ot">&lt;https://github.com/aertslab/pySCENIC&gt;</span>). The protocol is probably the best way to run this type of analysis currently: <span class="ot">&lt;https://github.com/aertslab/SCENICprotocol&gt;</span></span>
<span id="cb172-716"><a href="#cb172-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-717"><a href="#cb172-717" aria-hidden="true" tabindex="-1"></a><span class="fu">### CNV analysis</span></span>
<span id="cb172-718"><a href="#cb172-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-719"><a href="#cb172-719" aria-hidden="true" tabindex="-1"></a>Copy Number Variation (CNV) analysis is a computational approach used to infer genomic copy number alterations from transcriptomic data. It leverages the variability in gene expression levels across the genome to detect gains or losses of genomic regions.</span>
<span id="cb172-720"><a href="#cb172-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-721"><a href="#cb172-721" aria-hidden="true" tabindex="-1"></a>This method is particularly valuable in studies of cancer and other genetic disorders, where CNVs play a crucial role in disease progression and heterogeneity.</span>
<span id="cb172-722"><a href="#cb172-722" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-723"><a href="#cb172-723" aria-hidden="true" tabindex="-1"></a>The most well-known tool used for this analysis is inferCNV, which can be extremely slow and doesn't perform very well with cells with low-read count. This is why we implemented FastCNV, available to install for R on Github: <span class="ot">&lt;https://github.com/must-bioinfo/fastCNVdata&gt;</span></span>
<span id="cb172-724"><a href="#cb172-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-725"><a href="#cb172-725" aria-hidden="true" tabindex="-1"></a><span class="fu">### Ligand-receptor analysis</span></span>
<span id="cb172-726"><a href="#cb172-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-727"><a href="#cb172-727" aria-hidden="true" tabindex="-1"></a>Ligand-receptor analysis is a computational strategy to infer cell-cell communication by identifying interactions between signaling molecules (ligands) and their corresponding receptors expressed on different cell types or states.</span>
<span id="cb172-728"><a href="#cb172-728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-729"><a href="#cb172-729" aria-hidden="true" tabindex="-1"></a>This analysis involves systematically pairing ligands expressed by one or more cell populations with receptors expressed by neighboring or interacting cells, thereby predicting potential signaling pathways that facilitate intercellular communication.</span>
<span id="cb172-730"><a href="#cb172-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-731"><a href="#cb172-731" aria-hidden="true" tabindex="-1"></a>Well known tools are CellPhoneDB and CellChat, but LIANA+ provides wrappers and an all-in-one framework for this kind of analysis: <span class="ot">&lt;https://liana-py.readthedocs.io/en/latest/&gt;</span></span>
<span id="cb172-732"><a href="#cb172-732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-733"><a href="#cb172-733" aria-hidden="true" tabindex="-1"></a>Notably, this analysis can be much more powerful in spatially-resolved data.</span>
<span id="cb172-734"><a href="#cb172-734" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-735"><a href="#cb172-735" aria-hidden="true" tabindex="-1"></a><span class="fu">### Object conversion</span></span>
<span id="cb172-736"><a href="#cb172-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-737"><a href="#cb172-737" aria-hidden="true" tabindex="-1"></a>Format conversion is often necessary when working with single-cell RNA-seq to use</span>
<span id="cb172-738"><a href="#cb172-738" aria-hidden="true" tabindex="-1"></a>tools available for different objects.</span>
<span id="cb172-739"><a href="#cb172-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-740"><a href="#cb172-740" aria-hidden="true" tabindex="-1"></a>This can be easy between SingleCellExperiment and SeuratObject, as we used it</span>
<span id="cb172-741"><a href="#cb172-741" aria-hidden="true" tabindex="-1"></a>in this tutorial with <span class="in">`as.SingleCellExperiment`</span> and <span class="in">`as.Seurat`</span>.</span>
<span id="cb172-742"><a href="#cb172-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-743"><a href="#cb172-743" aria-hidden="true" tabindex="-1"></a>However, these functions sometimes don't work and the Seurat solution for h5</span>
<span id="cb172-744"><a href="#cb172-744" aria-hidden="true" tabindex="-1"></a>objects, SeuratDisk, fails its own tutorial. <span class="in">`sceasy`</span> (https://github.com/cellgeni/sceasy) </span>
<span id="cb172-745"><a href="#cb172-745" aria-hidden="true" tabindex="-1"></a>was made to resolve these conversion problems.</span>
<span id="cb172-746"><a href="#cb172-746" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-747"><a href="#cb172-747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-748"><a href="#cb172-748" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"loop":false,"selector":".lightbox","openEffect":"zoom","closeEffect":"zoom","descPosition":"bottom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>